{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "2020113002-IntroToBio-Assignment1-PhylogeneticAnalysis-UPGMA-Method.ipynb",
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3R8VEBVdKmIp"
      },
      "source": [
        "# Questions:\n",
        "\n",
        "1.\tConstruct a phylogenetic relationship for the given nucleotide sequences (Nucleotide.txt). \n",
        "\n",
        "  1.\tWrite a script (q1a) to generate a distance matrix csv file for the sequences present in the data file. Name the distance matrix file as 'Ndistance.txt'.\n",
        "\n",
        "* For example, \n",
        "\n",
        "* seq1 = 'ATGCATGCAA' \n",
        "\n",
        "* seq2 = 'ATGCATGCTA' \n",
        "\n",
        "* Distance (seq1, seq2) = Mismatches/total length = 1/10 = 0.1 \n",
        "\n",
        "  2.\tWrite a script (q1b) that uses 'Ndistance.txt' and generate phylogenetic relationship between the organisms using UPGMA method. \n",
        "\n",
        "\n",
        "2.\tConstruct a phylogenetic relationship for the given protein sequences (Protein.txt).\n",
        "\n",
        "  1.\tWrite a script (q2a) to generate a distance matrix csv file for the sequences present in the data file. Name the distance matrix file as 'Pdistance.txt'. Use BLOSUM62 for getting score values. \n",
        "\n",
        "  2.\tWrite a script (q2b) that uses 'Pdistance.txt' and generate phylogenetic relationship between the organisms using UPGMA method.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "I1Xev4vDB-ig"
      },
      "source": [
        "The Goal for this notebook is \n",
        "\n",
        "* ***Q1.*** To do phylogenetic tree analysis of the Nucleotide sequences provided.\n",
        "\n",
        "* ***Q2.*** To construct a phylogenetic relationship tree diagram for the protien seqeunces provided."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-e1zg4qPCETD"
      },
      "source": [
        "#***THOUGHTS AND NOTES:*** <u>UPGMA ASSIGNMENT</u>\n",
        "\n",
        "## <u>1. What even is UPGMA?</u>\n",
        "\n",
        "**UPGMA** - stands for <u>**U**</u>nweighted <u>**P**</u>air- <u>**G**</u>roup <u>**M**</u>ethod with <u>**A**</u>rithmetic mean.\n",
        "\n",
        "Thats a mouthful after unpacking, so lets make it simpler to grasp - \n",
        "\n",
        "* ***Unweighted*** - All pairwise distances contribute equally. <s>(There is an assumption here that needs to be adressed in any further attempts to this problem)</s>\n",
        "* ***Pair- Group*** - groups are combined in pairs (dichotomies only).\n",
        "* ***Arithmetic Mean*** - pairwise distances to each group (clade) are mean distances to all members of that group.\n",
        "\n",
        "### <u>a. *What was the assumption?*</u>\n",
        "\n",
        "UPGMA is a distance method and therefore needs a distance matrix. UPGMA is \"ultrametric\", meaning that all the terminal nodes (i.e. the sequences/taxa) are **equally distanced from the root**. In molecular terms, this means that UPGMA assumes a **molecular clock**, i.e. all lineages are evolving at a **constant rate**. In practical terms, this means that you can construct a **distance scale bar** and all the terminal nodes will be **level at position 0**.\n",
        "\n",
        "0, representing the present.\n",
        "\n",
        "## <u>2. How to do upgma ?</u>\n",
        "\n",
        "Each round of UPGMA follows the same pattern. \n",
        "1. Identify the shortest pairwise distance in the matrix. This identifies the two sequences to be clustered.\n",
        "2. Join the two sequences identified. \n",
        "3. The pair should be linked at a depth that is half of the shortest pairwise distance.\n",
        "4. The tip-to-tip distance between the joined elements will equal the shortest distance.\n",
        "\n",
        "## <u>3. Why UPGMA?</u>\n",
        "\n",
        "UPGMA is a method to tackle the problem of finding out which organism is most related to which one, given we have the genetic sequences of <u>**all**</u> said organisms. \n",
        "\n",
        "<u>**TLDR:**</u> Get genetic sequences as <u>*input*</u>, send Phylogenetic tree as <u>*Output*</u>. \n",
        "\n",
        "Great. now we have a rough idea of what exactly the beast we are tackling is. At least I do. I have referred to an example provided at http://www.nmsr.org/upgma.htm to help write this program better. "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ETgFAgV3DkZ2"
      },
      "source": [
        "# 4. Break it down - Divide and Conquer\n",
        "THIS IS A COMPLEX PROBLEM WITH LOTS OF MOVING PARTS.\n",
        "\n",
        "So Lets think of some simple steps to tackle this problem in.\n",
        "\n",
        "1. Get genetic sequences as input ( Needs to be tackled seperately for q1 and q2.)\n",
        "2. Align These sequences, and find the distance between each pair. (We haven't *exactly* been taught this, but we have been given a link to an online tool to do this) <s>(This problem is known to be **<i>N-P</i> HARD**) </s> I used biopython for this.\n",
        "3. Create a distance matrix out of these. ( should be easy enough, and simple to do for Q1 and Q2 ) easily done using csv\n",
        "4. Do the UPGMA algo on this.\n",
        "    \n",
        "  * this is gonna be involved, but should be same for both Q1 and Q2. \n",
        "  * yup this was involved, had to read a lot of blogs to do this. This newick form one was another pain to to. ho gaya bas ... \n",
        "5. Note the outputs and Construct a phylogenetic tree from these, and output in a given format.\n",
        "  * this is gonna be involved, but should be same for both Q1 and Q2. \n",
        "  * Used Newick form, sabse aasan tha, I swear 100% copy maara hai.\n",
        "  *  skipping the visualisation of newick as a visual tree, there are visualisers that take newick form and show a tree, please use one , I am not doing this, it is a data visualisation technique (newick to visual tree) that is not relevant to the UPGMA method. ( already exists online )\n",
        "\n",
        "We need 4 Script files, q1a.py, q1b.py, q2a.py, q2b.py."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "5_gMikLgaU5A"
      },
      "source": [
        "#Ensuring Dependencies : Run first when starting files check"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "enVfw5JhacV8",
        "outputId": "f6df6a94-9659-418d-e5f5-7a8af9fd0d3c"
      },
      "source": [
        "pip install biopython"
      ],
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: biopython in /usr/local/lib/python3.7/dist-packages (1.79)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython) (1.19.5)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "XxYA0Ik-bHP3",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 243
        },
        "outputId": "fabb8148-0754-423b-e8a9-7fa7151364dd"
      },
      "source": [
        "#Create Nucleotide.txt and Ndistance.txt to run q1a.py and q1b.py\n",
        "\n",
        "\n",
        "\n",
        "#Copy paste whatever nucleotide.txt should contain, use FASTA format.\n",
        "#Instead of running this code block you may also upload the file into the $PWD of google colab.\n",
        "\n",
        "nucleotideTextInput='''>NM_205222.3 Gallus gallus insulin (INS), mRNA\n",
        "ATATAAATATGGGAAAGAGAATGGGGAAATTTCTACCAGTCTTCATCTCTGAGAGCAAACTTCTCTGCAT\n",
        "CTCTTTCTCTCTTCTCTGGGCCTCCCCCAGCTCATCATGGCTCTCTGGATCCGATCACTGCCTCTTCTGG\n",
        "CTCTCCTTGTCTTTTCTGGCCCTGGAACCAGCTATGCAGCTGCCAACCAGCACCTCTGTGGCTCCCACTT\n",
        "GGTGGAGGCTCTCTACCTGGTGTGTGGAGAGCGTGGCTTCTTCTACTCCCCCAAAGCCCGACGGGATGTC\n",
        "GAGCAGCCCCTAGTGAGCAGTCCCTTGCGTGGCGAGGCAGGAGTGCTGCCTTTCCAGCAGGAGGAATACG\n",
        "AGAAAGTCAAGCGAGGGATTGTTGAGCAATGCTGCCATAACACGTGTTCCCTCTACCAACTGGAGAACTA\n",
        "CTGCAACTAGCCAAGAAGCCGGAAGCGGGCACAGACATACACTTACTCTATCGCACCTTCAAAGCATTTG\n",
        "AATAAACCTTGTTGGTCTACTGGAAGACTTGTGCC\n",
        "\n",
        ">NM_001130093.2 Canis lupus familiaris insulin (INS), mRNA\n",
        "CACCCCGACACGGCCGGCAAACAGGTCGCCATGGCCCTCTGGATGCGCCTCCTGCCCCTGCTGGCCCTGC\n",
        "TGGCCCTCTGGGCGCCCGCGCCCACCCGAGCCTTCGTTAACCAGCACCTGTGTGGCTCCCACCTGGTAGA\n",
        "GGCTCTGTACCTGGTGTGCGGGGAGCGCGGCTTCTTCTACACGCCTAAGGCCCGCCGGGAGGTGGAGGAC\n",
        "CTGCAGGTGAGGGACGTGGAGCTGGCCGGGGCGCCTGGCGAGGGCGGCCTGCAGCCCCTGGCCCTGGAGG\n",
        "GGGCCCTGCAGAAGCGAGGCATCGTGGAGCAGTGCTGCACCAGCATCTGCTCCCTCTACCAGCTGGAGAA\n",
        "TTACTGCAACTAGGGGCGCGGGGGGCAGGACGTGGCAGCACCTGCTGCAGGTCACGGTGGCCGCAAGCCT\n",
        "TCGGCTCTCTGCACCCCAAGTGATTCAATAAACCCTCTGAATG\n",
        "\n",
        ">KX951416.1 Labeo rohita insulin mRNA, complete cds\n",
        "CCGATGTGCTCTGAAAGCCTGGATGCAAAAACACCTTCTCTTGTCTACCATCTCTACCATTCCTTGTCCT\n",
        "CTGCTGCAAGAACAGTGTGACCATGGCAGTGTGGCTCCAGGCTGGTGCTCTTTTGTTCTTGTTGGCCGTC\n",
        "TCCAGTGTGAACGCTAACGCAGGGGCCCCACAGCATCTGTGTGGATCTCATCTGGTCGATGCCCTCTACC\n",
        "TGGTCTGTGGTCCAACAGGATTCTTCTACAACCCCAAGAGAGATGTTGACCCTCTTATGGGTTTCCTTCC\n",
        "TCCAAAATCTGCCCAGGAAACTGAGGTGGCTGACTTTGCATTTAAAGATCATGCCGAGGTGATAAGGAAG\n",
        "AGAGGCATTGTGGAGCAGTGTTGCCACAAACCCTGCAGTATCTTTGAGCTGCAGAACTACTGTAACTAAA\n",
        "GAACCTGCACATGTCTTGTGACAACTGCCAGTGACTTTACCACCTGTTTGCACACAGGTATCATGGGAAT\n",
        "TACGAGAACTAAACAAAAAGTATCTTTTATTTTAAAATAGTTTACTTTT\n",
        "\n",
        ">BT006808.1 Homo sapiens insulin mRNA, complete cds\n",
        "ATGGCCCTGTGGATGCGCCTCCTGCCCCTGCTGGCGCTGCTGGCCCTCTGGGGACCTGACCCAGCCGCAG\n",
        "CCTTTGTGAACCAACACCTGTGCGGCTCACACCTGGTGGAAGCTCTCTACCTAGTGTGCGGGGAACGAGG\n",
        "CTTCTTCTACACACCCAAGACCCGCCGGGAGGCAGAGGACCTGCAGGTGGGGCAGGTGGAGCTGGGCGGG\n",
        "GGCCCTGGTGCAGGCAGCCTGCAGCCCTTGGCCCTGGAGGGGTCCCTGCAGAAGCGTGGCATTGTGGAAC\n",
        "AATGCTGTACCAGCATCTGCTCCCTCTACCAGCTGGAGAACTACTGCAACTAG\n",
        "\n",
        ">NM_019129.3 Rattus norvegicus insulin 1 (Ins1), mRNA\n",
        "AACCCTAAGTGACCAGCTACAATCATAGACCATCAGCAAGCAGGTCATTGTTCCAACATGGCCCTGTGGA\n",
        "TGCGCTTCCTGCCCCTGCTGGCCCTGCTCGTCCTCTGGGAGCCCAAGCCTGCCCAGGCTTTTGTCAAACA\n",
        "GCACCTTTGTGGTCCTCACCTGGTGGAGGCTCTGTACCTGGTGTGTGGGGAACGTGGTTTCTTCTACACA\n",
        "CCCAAGTCCCGTCGTGAAGTGGAGGACCCGCAAGTGCCACAACTGGAGCTGGGTGGAGGCCCGGAGGCCG\n",
        "GGGATCTTCAGACCTTGGCACTGGAGGTTGCCCGGCAGAAGCGTGGCATTGTGGATCAGTGCTGCACCAG\n",
        "CATCTGCTCCCTCTACCAACTGGAGAACTACTGCAACTGAGTCCACCACTCCCCGCCCACCCCTCTGCAA\n",
        "TGAATAAAGCCTTTGAATGAGCACCAAAAAAAAAAAAAAAAAA\n",
        "\n",
        ">NM_001109772.1 Sus scrofa insulin (INS), mRNA\n",
        "AGCCCTCTGGGACCAGCTGTGTTCCCAGGCCACCGGCAAGCAGGTCCTCACCCCCCGCCATGGCCCTGTG\n",
        "GACGCGCCTCCTGCCCCTGCTGGCCCTGCTGGCCCTCTGGGCGCCCGCCCCGGCCCAGGCCTTCGTGAAC\n",
        "CAGCACCTGTGCGGCTCCCACCTGGTGGAGGCGCTGTACCTGGTGTGCGGGGAGCGCGGCTTCTTCTACA\n",
        "CGCCCAAGGCCCGTCGGGAGGCGGAGAACCCTCAGGCAGGTGCCGTGGAGCTGGGCGGAGGCCTGGGCGG\n",
        "CCTGCAGGCCCTGGCGCTGGAGGGGCCCCCGCAGAAGCGTGGCATCGTGGAGCAGTGCTGCACCAGCATC\n",
        "TGTTCCCTCTACCAGCTGGAGAACTACTGCAACTAGGCCGCCCCTGAGGGCGCCTGCTGCTCCCCGCACC\n",
        "CCAAAACCCAATAAA\n",
        "\n",
        ">NM_173926.2 Bos taurus insulin (INS), transcript variant 1, mRNA\n",
        "AGCCCCCCGCCCTCAGGACCGGCTGCATTCGAGGCTGCCAGCAAGCAGGTCCTCGCAGCCCCGCCATGGC\n",
        "CCTGTGGACACGCCTGGCGCCCCTGCTGGCCCTGCTGGCGCTCTGGGCCCCCGCCCCGGCCCGCGCCTTC\n",
        "GTCAACCAGCATCTGTGTGGCTCCCACCTGGTGGAGGCGCTGTACCTGGTGTGCGGAGAGCGCGGCTTCT\n",
        "TCTACACGCCCAAGGCCCGCCGGGAGGTGGAGGGCCCCCAGGTGGGGGCGCTGGAGCTGGCCGGAGGCCC\n",
        "GGGCGCGGGCGGCCTGGAGGGGCCCCCGCAGAAGCGTGGCATCGTGGAGCAGTGCTGTGCCAGCGTCTGC\n",
        "TCGCTCTACCAGCTGGAGAACTACTGTAACTAGGCCTGCCCCCGACACCAATAAACCCCTTGACGAGCCC\n",
        "TGCAAAAAAAAAAA\n",
        "\n",
        ">NM_001008996.2 Pan troglodytes insulin (INS), mRNA\n",
        "AGCCCTCCAGGACAGGCTGCATCAGAAGAGGCCATCAAGCAGATCACTGTCCTTCTGCCATGGCCCTGTG\n",
        "GATGCGCCTCCTGCCCCTGCTGGTGCTGCTGGCCCTCTGGGGACCTGACCCAGCCTCGGCCTTTGTGAAC\n",
        "CAACACCTGTGCGGCTCCCACCTGGTGGAAGCTCTCTACCTAGTGTGCGGGGAACGAGGCTTCTTCTACA\n",
        "CACCCAAGACCCGCCGGGAGGCAGAGGACCTGCAGGTGGGGCAGGTGGAGCTGGGCGGGGGCCCTGGTGC\n",
        "AGGCAGCCTGCAGCCCTTGGCCCTGGAGGGGTCCCTGCAGAAGCGTGGTATCGTGGAACAATGCTGTACC\n",
        "AGCATCTGCTCCCTCTACCAGCTGGAGAACTACTGCAACTAGATGGAATAAAGCCCTTGAACCAGC\n",
        "\n",
        ">M57671.1 Octodon degus insulin mRNA, complete cds\n",
        "GCATTCTGAGGCATTCTCTAACAGGTTCTCGACCCTCCGCCATGGCCCCGTGGATGCATCTCCTCACCGT\n",
        "GCTGGCCCTGCTGGCCCTCTGGGGACCCAACTCTGTTCAGGCCTATTCCAGCCAGCACCTGTGCGGCTCC\n",
        "AACCTAGTGGAGGCACTGTACATGACATGTGGACGGAGTGGCTTCTATAGACCCCACGACCGCCGAGAGC\n",
        "TGGAGGACCTCCAGGTGGAGCAGGCAGAACTGGGTCTGGAGGCAGGCGGCCTGCAGCCTTCGGCCCTGGA\n",
        "GATGATTCTGCAGAAGCGCGGCATTGTGGATCAGTGCTGTAATAACATTTGCACATTTAACCAGCTGCAG\n",
        "AACTACTGCAATGTCCCTTAGACACCTGCCTTGGGCCTGGCCTGCTGCTCTGCCCTGGCAACCAATAAAC\n",
        "CCCTTGAATGAG\n",
        "\n",
        ">U03610.1 Oryctolagus cuniculus New Zealand White insulin mRNA, complete cds\n",
        "TCATCGGCTCTGCACCATGGCCTCCCTGGCCGCGCTCCTGCCCCTGCTGGCCCTGCTGGTCCTCTGCAGA\n",
        "CTGGATCCTGCCCAGGCCTTCGTCAACCAGCACCTGTGCGGCTCTCACCTGGTGGAGGCGCTGTACCTGG\n",
        "TGTGCGGGGAGCGCGGCTTTTTTTATACACCCAAGTCCCGCCGCGAGGTGGAGGAGCTGCAGGTGGGGCA\n",
        "GGCGGAGCTGGGCGGGGGGCCCGGCGCGGGCGGCCTGCAGCCCTCGGCGCTGGAGCTGGCCCTGCAGAAG\n",
        "CGCGGCATCGTGGAGCAGTGTTGCACCAGCATCTGCTCGCTCTACCAGCTGGAGAACTACTGCAACTAGG\n",
        "GGTGCCCCCCACCCACCCCTGCCCGCGCCCCCCACGCCCCCCGCCCTCGCCCCCACCCAATAAACCCCTC\n",
        "CACGCGCCCCGGG\n",
        "\n",
        "\n",
        "'''\n",
        "\n",
        "with open(\"Nucleotide.txt\", 'x') as f_out:\n",
        "  f_out.write( nucleotideTextInput )\n",
        "\n",
        "with open(\"Ndistance.txt\",'x') as f_out:\n",
        "  f_out.write('')\n",
        "\n",
        "exit(0)\n",
        "\n"
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "error",
          "ename": "FileExistsError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFileExistsError\u001b[0m                           Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-1-45b0900c3e26>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     95\u001b[0m '''\n\u001b[1;32m     96\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 97\u001b[0;31m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Nucleotide.txt\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'x'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf_out\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     98\u001b[0m   \u001b[0mf_out\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0mnucleotideTextInput\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFileExistsError\u001b[0m: [Errno 17] File exists: 'Nucleotide.txt'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dRHngM_0Bfuz",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "109b5dcc-6ba0-46ec-b77d-d5ded98fe3f9"
      },
      "source": [
        "\n",
        "#q1a.py\n",
        "\n",
        "'''\n",
        " After reading given input of Nucleotide.txt, I see that each new entry begins with a '>',\n",
        " followed by a line of Information on the gene, followed by some N lines of ATGC\n",
        "\n",
        " Identified as the FASTA format.\n",
        "\n",
        " I shall try to use biopython to give us the (n^2 - n)/2 values for distance matrix.\n",
        "\n",
        " Resources referred to \n",
        " \n",
        " 1. https://python.omics.wiki/biopython/examples/read-fasta\n",
        " 2. https://www.geeksforgeeks.org/biopython-pairwise-alignment/\n",
        " 3. https://stackoverflow.com/questions/20580657/how-to-read-a-fasta-file-in-python \n",
        " 4. https://vlab.amrita.edu/index.php?sub=3&brch=274&sim=1438&cnt=1 \n",
        " 5. https://www.ebi.ac.uk/seqdb/confluence/display/JDSAT/Clustal+Omega+Help+and+Documentation#ClustalOmegaHelpandDocumentation-sequence \n",
        " 6. https://www.tutorialspoint.com/biopython/biopython_sequence_alignments.htm \n",
        " 7. https://www.geeksforgeeks.org/biopython-sequence-alignment/#:~:text=Biopython%20%E2%80%93%20Sequence%20Alignment%20Last%20Updated%20%3A%2011,identify%20the%20region%20of%20similarity%20among%20them.%20 \n",
        " 8. https://www.ncbi.nlm.nih.gov/sra/docs/submitformats/#:~:text=fasta%20files%20may%20be%20submitted%20with%20corresponding%20qual%20files%2C%20too.%20these%20are%20recognized%20in%20the%20sra%20data%20processing%20pipeline%20as%20equivalent%20to%20fastq%20and%20should%20be%20specified%20as%20fastq%20when%20submitting%20the%20data%20files. \n",
        " 9. https://www.csestack.org/python-check-if-file-directory-exists/\n",
        " 10. https://flexiple.com/python-create-file/ \n",
        " 11. https://www.askpython.com/python/built-in-methods/python-print-to-file#:~:text=Python%20%E2%80%93%20Print%20to%20File%201%20Method%201%3A,4%20Method%204%3A%20Use%20the%20logging%20module.%20 \n",
        " 12. https://stackoverflow.com/questions/9426045/difference-between-exit0-and-exit1-in-python#:~:text=exit%20%281%29%20means%20there%20was%20some%20issue%20%2F,A%20zero%20error%20code%20means%20a%20successful%20exit. \n",
        "\n",
        "\n",
        "'''\n",
        "\n",
        "import csv\n",
        "import os\n",
        "from Bio import SeqIO\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from Bio.Seq import Seq\n",
        "from Bio import pairwise2\n",
        "\n",
        "def merge(list1, list2):\n",
        "      \n",
        "    merged_list = [(list1[i], list2[i]) for i in range(0, len(list1))]\n",
        "    return merged_list\n",
        "\n",
        "# Python program to get transpose\n",
        "# elements of two dimension list\n",
        "def transpose(l1, l2):\n",
        " \n",
        "    # iterate over list l1 to the length of an item\n",
        "    for i in range(len(l1[0])):\n",
        "        # print(i)\n",
        "        row =[]\n",
        "        for item in l1:\n",
        "            # appending to new list with values and index positions\n",
        "            # i contains index position and item contains values\n",
        "            row.append(item[i])\n",
        "        l2.append(row)\n",
        "    return l2\n",
        "\n",
        "def main ():\n",
        "#to help resolve indent errors \n",
        "    file_in ='./Nucleotide.txt'\n",
        "    file_out='./Ndistance.txt'\n",
        "\n",
        "    if os.access(file_in, os.R_OK):\n",
        "        print (\"File \", file_in ,\" is accessible to read\")\n",
        "    else:\n",
        "        print (\"File \", file_in ,\" is not accessible to read, exiting.\")\n",
        "        exit (1)\n",
        "    \n",
        "    if os.access(file_out, os.W_OK):\n",
        "        print (\"File \", file_out,\" is accessible to write \\n\")\n",
        "    else:\n",
        "        print (\"File \", file_out,\" is not accessible to write, exiting.\")\n",
        "        exit (1)\n",
        "\n",
        "    IdList = list ()\n",
        "    NucleotideSeqList= list ()\n",
        "\n",
        "    with open(file_out, 'w') as f_out:\n",
        "    #to help resolve indent errors   \n",
        "        print(\"Beginning to Read \",file_in,\"\\n\")\n",
        "        for seq_record in SeqIO.parse(open(file_in, mode='r'), 'fasta'):\n",
        "        #to help resolve indent errors \n",
        "\n",
        "            # remove .id from .description record (remove all before first space)\n",
        "            seq_record.description=' '.join(seq_record.description.split()[1:])\n",
        "            \n",
        "            # do something (print or edit seq_record)       \n",
        "            print('SequenceID = '  + seq_record.id)\n",
        "            print('Description = ' + seq_record.description)\n",
        "            print('Sequence = ' + seq_record.seq + '\\n' )\n",
        "\n",
        "            # Save the Sequence into NucleotideSeqList and Id into IdList\n",
        "            NucleotideSeqList.append( seq_record.seq )\n",
        "            IdList.append( seq_record.id )\n",
        "\n",
        "        no_of_nucleotides = len(NucleotideSeqList )\n",
        "        print(no_of_nucleotides,'Nucleotides scanned into our List\\n')\n",
        "        print(\"Our Lists are : \",merge ( IdList, NucleotideSeqList ))\n",
        "        print(NucleotideSeqList)\n",
        "        print(IdList,\"\\n\")\n",
        "\n",
        "        #Initialize distance matrix\n",
        "        dist_Matrix=list()\n",
        "\n",
        "        #Now to find distances, pairwise \n",
        "        for i in range(0,no_of_nucleotides,1):\n",
        "            column= list ()\n",
        "            for i_1 in range(0,i+1,1):\n",
        "                column.append(0)\n",
        "            for j in range (i+1, no_of_nucleotides, 1):\n",
        "                #compare with j'th nucleotide, and find best pairing.\n",
        "                alignments = pairwise2.align.globalxx(NucleotideSeqList[i], NucleotideSeqList[j])\n",
        "                length = ( max(len(NucleotideSeqList[i]), len(NucleotideSeqList[j])) // 1)\n",
        "                score= 0\n",
        "                for alignment in alignments: \n",
        "                    # print( (alignment[2]//1) , (alignment[4]//1) , ( (alignment[2]//1)/ (alignment[4]//1) ) ) \n",
        "                    if ( (alignment[2]//1) >  score ):\n",
        "                        score = (alignment[2]//1)\n",
        "                norm_score = score/length\n",
        "                print(score,length,norm_score)\n",
        "                #add to 'column' IMPORTANT LINE - CHANGE THE \"score\" to \"norm_score\" to get normalized ans.\n",
        "                column.append(norm_score)\n",
        "            dist_Matrix.append(column)\n",
        "        out_Matrix=list ()\n",
        "        out_Matrix = transpose (dist_Matrix, out_Matrix)\n",
        "        print(out_Matrix)\n",
        "        writer = csv.writer(f_out)\n",
        "        # f_out.write(str (no_of_nucleotides)+\",\\n\")\n",
        "        writer.writerows(out_Matrix)\n",
        "    file_out_2=\"./Ntaxa.txt\"\n",
        "    with open(file_out_2, 'w') as f_out_2:\n",
        "      writer = csv.writer(f_out_2)\n",
        "      writer.writerow (IdList)\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    main()\n",
        "\n"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "File  ./Nucleotide.txt  is accessible to read\n",
            "File  ./Ndistance.txt  is accessible to write \n",
            "\n",
            "Beginning to Read  ./Nucleotide.txt \n",
            "\n",
            "SequenceID = NM_205222.3\n",
            "Description = Gallus gallus insulin (INS), mRNA\n",
            "Sequence = ATATAAATATGGGAAAGAGAATGGGGAAATTTCTACCAGTCTTCATCTCTGAGAGCAAACTTCTCTGCATCTCTTTCTCTCTTCTCTGGGCCTCCCCCAGCTCATCATGGCTCTCTGGATCCGATCACTGCCTCTTCTGGCTCTCCTTGTCTTTTCTGGCCCTGGAACCAGCTATGCAGCTGCCAACCAGCACCTCTGTGGCTCCCACTTGGTGGAGGCTCTCTACCTGGTGTGTGGAGAGCGTGGCTTCTTCTACTCCCCCAAAGCCCGACGGGATGTCGAGCAGCCCCTAGTGAGCAGTCCCTTGCGTGGCGAGGCAGGAGTGCTGCCTTTCCAGCAGGAGGAATACGAGAAAGTCAAGCGAGGGATTGTTGAGCAATGCTGCCATAACACGTGTTCCCTCTACCAACTGGAGAACTACTGCAACTAGCCAAGAAGCCGGAAGCGGGCACAGACATACACTTACTCTATCGCACCTTCAAAGCATTTGAATAAACCTTGTTGGTCTACTGGAAGACTTGTGCC\n",
            "\n",
            "SequenceID = NM_001130093.2\n",
            "Description = Canis lupus familiaris insulin (INS), mRNA\n",
            "Sequence = CACCCCGACACGGCCGGCAAACAGGTCGCCATGGCCCTCTGGATGCGCCTCCTGCCCCTGCTGGCCCTGCTGGCCCTCTGGGCGCCCGCGCCCACCCGAGCCTTCGTTAACCAGCACCTGTGTGGCTCCCACCTGGTAGAGGCTCTGTACCTGGTGTGCGGGGAGCGCGGCTTCTTCTACACGCCTAAGGCCCGCCGGGAGGTGGAGGACCTGCAGGTGAGGGACGTGGAGCTGGCCGGGGCGCCTGGCGAGGGCGGCCTGCAGCCCCTGGCCCTGGAGGGGGCCCTGCAGAAGCGAGGCATCGTGGAGCAGTGCTGCACCAGCATCTGCTCCCTCTACCAGCTGGAGAATTACTGCAACTAGGGGCGCGGGGGGCAGGACGTGGCAGCACCTGCTGCAGGTCACGGTGGCCGCAAGCCTTCGGCTCTCTGCACCCCAAGTGATTCAATAAACCCTCTGAATG\n",
            "\n",
            "SequenceID = KX951416.1\n",
            "Description = Labeo rohita insulin mRNA, complete cds\n",
            "Sequence = CCGATGTGCTCTGAAAGCCTGGATGCAAAAACACCTTCTCTTGTCTACCATCTCTACCATTCCTTGTCCTCTGCTGCAAGAACAGTGTGACCATGGCAGTGTGGCTCCAGGCTGGTGCTCTTTTGTTCTTGTTGGCCGTCTCCAGTGTGAACGCTAACGCAGGGGCCCCACAGCATCTGTGTGGATCTCATCTGGTCGATGCCCTCTACCTGGTCTGTGGTCCAACAGGATTCTTCTACAACCCCAAGAGAGATGTTGACCCTCTTATGGGTTTCCTTCCTCCAAAATCTGCCCAGGAAACTGAGGTGGCTGACTTTGCATTTAAAGATCATGCCGAGGTGATAAGGAAGAGAGGCATTGTGGAGCAGTGTTGCCACAAACCCTGCAGTATCTTTGAGCTGCAGAACTACTGTAACTAAAGAACCTGCACATGTCTTGTGACAACTGCCAGTGACTTTACCACCTGTTTGCACACAGGTATCATGGGAATTACGAGAACTAAACAAAAAGTATCTTTTATTTTAAAATAGTTTACTTTT\n",
            "\n",
            "SequenceID = BT006808.1\n",
            "Description = Homo sapiens insulin mRNA, complete cds\n",
            "Sequence = ATGGCCCTGTGGATGCGCCTCCTGCCCCTGCTGGCGCTGCTGGCCCTCTGGGGACCTGACCCAGCCGCAGCCTTTGTGAACCAACACCTGTGCGGCTCACACCTGGTGGAAGCTCTCTACCTAGTGTGCGGGGAACGAGGCTTCTTCTACACACCCAAGACCCGCCGGGAGGCAGAGGACCTGCAGGTGGGGCAGGTGGAGCTGGGCGGGGGCCCTGGTGCAGGCAGCCTGCAGCCCTTGGCCCTGGAGGGGTCCCTGCAGAAGCGTGGCATTGTGGAACAATGCTGTACCAGCATCTGCTCCCTCTACCAGCTGGAGAACTACTGCAACTAG\n",
            "\n",
            "SequenceID = NM_019129.3\n",
            "Description = Rattus norvegicus insulin 1 (Ins1), mRNA\n",
            "Sequence = AACCCTAAGTGACCAGCTACAATCATAGACCATCAGCAAGCAGGTCATTGTTCCAACATGGCCCTGTGGATGCGCTTCCTGCCCCTGCTGGCCCTGCTCGTCCTCTGGGAGCCCAAGCCTGCCCAGGCTTTTGTCAAACAGCACCTTTGTGGTCCTCACCTGGTGGAGGCTCTGTACCTGGTGTGTGGGGAACGTGGTTTCTTCTACACACCCAAGTCCCGTCGTGAAGTGGAGGACCCGCAAGTGCCACAACTGGAGCTGGGTGGAGGCCCGGAGGCCGGGGATCTTCAGACCTTGGCACTGGAGGTTGCCCGGCAGAAGCGTGGCATTGTGGATCAGTGCTGCACCAGCATCTGCTCCCTCTACCAACTGGAGAACTACTGCAACTGAGTCCACCACTCCCCGCCCACCCCTCTGCAATGAATAAAGCCTTTGAATGAGCACCAAAAAAAAAAAAAAAAAA\n",
            "\n",
            "SequenceID = NM_001109772.1\n",
            "Description = Sus scrofa insulin (INS), mRNA\n",
            "Sequence = AGCCCTCTGGGACCAGCTGTGTTCCCAGGCCACCGGCAAGCAGGTCCTCACCCCCCGCCATGGCCCTGTGGACGCGCCTCCTGCCCCTGCTGGCCCTGCTGGCCCTCTGGGCGCCCGCCCCGGCCCAGGCCTTCGTGAACCAGCACCTGTGCGGCTCCCACCTGGTGGAGGCGCTGTACCTGGTGTGCGGGGAGCGCGGCTTCTTCTACACGCCCAAGGCCCGTCGGGAGGCGGAGAACCCTCAGGCAGGTGCCGTGGAGCTGGGCGGAGGCCTGGGCGGCCTGCAGGCCCTGGCGCTGGAGGGGCCCCCGCAGAAGCGTGGCATCGTGGAGCAGTGCTGCACCAGCATCTGTTCCCTCTACCAGCTGGAGAACTACTGCAACTAGGCCGCCCCTGAGGGCGCCTGCTGCTCCCCGCACCCCAAAACCCAATAAA\n",
            "\n",
            "SequenceID = NM_173926.2\n",
            "Description = Bos taurus insulin (INS), transcript variant 1, mRNA\n",
            "Sequence = AGCCCCCCGCCCTCAGGACCGGCTGCATTCGAGGCTGCCAGCAAGCAGGTCCTCGCAGCCCCGCCATGGCCCTGTGGACACGCCTGGCGCCCCTGCTGGCCCTGCTGGCGCTCTGGGCCCCCGCCCCGGCCCGCGCCTTCGTCAACCAGCATCTGTGTGGCTCCCACCTGGTGGAGGCGCTGTACCTGGTGTGCGGAGAGCGCGGCTTCTTCTACACGCCCAAGGCCCGCCGGGAGGTGGAGGGCCCCCAGGTGGGGGCGCTGGAGCTGGCCGGAGGCCCGGGCGCGGGCGGCCTGGAGGGGCCCCCGCAGAAGCGTGGCATCGTGGAGCAGTGCTGTGCCAGCGTCTGCTCGCTCTACCAGCTGGAGAACTACTGTAACTAGGCCTGCCCCCGACACCAATAAACCCCTTGACGAGCCCTGCAAAAAAAAAAA\n",
            "\n",
            "SequenceID = NM_001008996.2\n",
            "Description = Pan troglodytes insulin (INS), mRNA\n",
            "Sequence = AGCCCTCCAGGACAGGCTGCATCAGAAGAGGCCATCAAGCAGATCACTGTCCTTCTGCCATGGCCCTGTGGATGCGCCTCCTGCCCCTGCTGGTGCTGCTGGCCCTCTGGGGACCTGACCCAGCCTCGGCCTTTGTGAACCAACACCTGTGCGGCTCCCACCTGGTGGAAGCTCTCTACCTAGTGTGCGGGGAACGAGGCTTCTTCTACACACCCAAGACCCGCCGGGAGGCAGAGGACCTGCAGGTGGGGCAGGTGGAGCTGGGCGGGGGCCCTGGTGCAGGCAGCCTGCAGCCCTTGGCCCTGGAGGGGTCCCTGCAGAAGCGTGGTATCGTGGAACAATGCTGTACCAGCATCTGCTCCCTCTACCAGCTGGAGAACTACTGCAACTAGATGGAATAAAGCCCTTGAACCAGC\n",
            "\n",
            "SequenceID = M57671.1\n",
            "Description = Octodon degus insulin mRNA, complete cds\n",
            "Sequence = GCATTCTGAGGCATTCTCTAACAGGTTCTCGACCCTCCGCCATGGCCCCGTGGATGCATCTCCTCACCGTGCTGGCCCTGCTGGCCCTCTGGGGACCCAACTCTGTTCAGGCCTATTCCAGCCAGCACCTGTGCGGCTCCAACCTAGTGGAGGCACTGTACATGACATGTGGACGGAGTGGCTTCTATAGACCCCACGACCGCCGAGAGCTGGAGGACCTCCAGGTGGAGCAGGCAGAACTGGGTCTGGAGGCAGGCGGCCTGCAGCCTTCGGCCCTGGAGATGATTCTGCAGAAGCGCGGCATTGTGGATCAGTGCTGTAATAACATTTGCACATTTAACCAGCTGCAGAACTACTGCAATGTCCCTTAGACACCTGCCTTGGGCCTGGCCTGCTGCTCTGCCCTGGCAACCAATAAACCCCTTGAATGAG\n",
            "\n",
            "SequenceID = U03610.1\n",
            "Description = Oryctolagus cuniculus New Zealand White insulin mRNA, complete cds\n",
            "Sequence = TCATCGGCTCTGCACCATGGCCTCCCTGGCCGCGCTCCTGCCCCTGCTGGCCCTGCTGGTCCTCTGCAGACTGGATCCTGCCCAGGCCTTCGTCAACCAGCACCTGTGCGGCTCTCACCTGGTGGAGGCGCTGTACCTGGTGTGCGGGGAGCGCGGCTTTTTTTATACACCCAAGTCCCGCCGCGAGGTGGAGGAGCTGCAGGTGGGGCAGGCGGAGCTGGGCGGGGGGCCCGGCGCGGGCGGCCTGCAGCCCTCGGCGCTGGAGCTGGCCCTGCAGAAGCGCGGCATCGTGGAGCAGTGTTGCACCAGCATCTGCTCGCTCTACCAGCTGGAGAACTACTGCAACTAGGGGTGCCCCCCACCCACCCCTGCCCGCGCCCCCCACGCCCCCCGCCCTCGCCCCCACCCAATAAACCCCTCCACGCGCCCCGGG\n",
            "\n",
            "10 Nucleotides scanned into our List\n",
            "\n",
            "Our Lists are :  [('NM_205222.3', Seq('ATATAAATATGGGAAAGAGAATGGGGAAATTTCTACCAGTCTTCATCTCTGAGA...GCC')), ('NM_001130093.2', Seq('CACCCCGACACGGCCGGCAAACAGGTCGCCATGGCCCTCTGGATGCGCCTCCTG...ATG')), ('KX951416.1', Seq('CCGATGTGCTCTGAAAGCCTGGATGCAAAAACACCTTCTCTTGTCTACCATCTC...TTT')), ('BT006808.1', Seq('ATGGCCCTGTGGATGCGCCTCCTGCCCCTGCTGGCGCTGCTGGCCCTCTGGGGA...TAG')), ('NM_019129.3', Seq('AACCCTAAGTGACCAGCTACAATCATAGACCATCAGCAAGCAGGTCATTGTTCC...AAA')), ('NM_001109772.1', Seq('AGCCCTCTGGGACCAGCTGTGTTCCCAGGCCACCGGCAAGCAGGTCCTCACCCC...AAA')), ('NM_173926.2', Seq('AGCCCCCCGCCCTCAGGACCGGCTGCATTCGAGGCTGCCAGCAAGCAGGTCCTC...AAA')), ('NM_001008996.2', Seq('AGCCCTCCAGGACAGGCTGCATCAGAAGAGGCCATCAAGCAGATCACTGTCCTT...AGC')), ('M57671.1', Seq('GCATTCTGAGGCATTCTCTAACAGGTTCTCGACCCTCCGCCATGGCCCCGTGGA...GAG')), ('U03610.1', Seq('TCATCGGCTCTGCACCATGGCCTCCCTGGCCGCGCTCCTGCCCCTGCTGGCCCT...GGG'))]\n",
            "[Seq('ATATAAATATGGGAAAGAGAATGGGGAAATTTCTACCAGTCTTCATCTCTGAGA...GCC'), Seq('CACCCCGACACGGCCGGCAAACAGGTCGCCATGGCCCTCTGGATGCGCCTCCTG...ATG'), Seq('CCGATGTGCTCTGAAAGCCTGGATGCAAAAACACCTTCTCTTGTCTACCATCTC...TTT'), Seq('ATGGCCCTGTGGATGCGCCTCCTGCCCCTGCTGGCGCTGCTGGCCCTCTGGGGA...TAG'), Seq('AACCCTAAGTGACCAGCTACAATCATAGACCATCAGCAAGCAGGTCATTGTTCC...AAA'), Seq('AGCCCTCTGGGACCAGCTGTGTTCCCAGGCCACCGGCAAGCAGGTCCTCACCCC...AAA'), Seq('AGCCCCCCGCCCTCAGGACCGGCTGCATTCGAGGCTGCCAGCAAGCAGGTCCTC...AAA'), Seq('AGCCCTCCAGGACAGGCTGCATCAGAAGAGGCCATCAAGCAGATCACTGTCCTT...AGC'), Seq('GCATTCTGAGGCATTCTCTAACAGGTTCTCGACCCTCCGCCATGGCCCCGTGGA...GAG'), Seq('TCATCGGCTCTGCACCATGGCCTCCCTGGCCGCGCTCCTGCCCCTGCTGGCCCT...GGG')]\n",
            "['NM_205222.3', 'NM_001130093.2', 'KX951416.1', 'BT006808.1', 'NM_019129.3', 'NM_001109772.1', 'NM_173926.2', 'NM_001008996.2', 'M57671.1', 'U03610.1'] \n",
            "\n",
            "341.0 525 0.6495238095238095\n",
            "362.0 539 0.6716141001855288\n",
            "272.0 525 0.518095238095238\n",
            "343.0 525 0.6533333333333333\n",
            "336.0 525 0.64\n",
            "324.0 525 0.6171428571428571\n",
            "316.0 525 0.6019047619047619\n",
            "327.0 525 0.6228571428571429\n",
            "313.0 525 0.5961904761904762\n",
            "316.0 539 0.5862708719851577\n",
            "296.0 463 0.6393088552915767\n",
            "352.0 463 0.7602591792656588\n",
            "373.0 463 0.8056155507559395\n",
            "356.0 463 0.7688984881209503\n",
            "346.0 463 0.7473002159827213\n",
            "332.0 463 0.7170626349892009\n",
            "360.0 463 0.7775377969762419\n",
            "271.0 539 0.5027829313543599\n",
            "327.0 539 0.6066790352504638\n",
            "305.0 539 0.5658627087198516\n",
            "312.0 539 0.5788497217068646\n",
            "306.0 539 0.5677179962894249\n",
            "315.0 539 0.5844155844155844\n",
            "291.0 539 0.5398886827458256\n",
            "283.0 463 0.6112311015118791\n",
            "292.0 435 0.671264367816092\n",
            "280.0 434 0.6451612903225806\n",
            "327.0 416 0.7860576923076923\n",
            "271.0 432 0.6273148148148148\n",
            "289.0 433 0.6674364896073903\n",
            "351.0 463 0.7580993520518359\n",
            "347.0 463 0.7494600431965442\n",
            "342.0 463 0.7386609071274298\n",
            "328.0 463 0.7084233261339092\n",
            "337.0 463 0.7278617710583153\n",
            "372.0 435 0.8551724137931035\n",
            "348.0 435 0.8\n",
            "330.0 435 0.7586206896551724\n",
            "340.0 435 0.7816091954022989\n",
            "341.0 434 0.7857142857142857\n",
            "315.0 434 0.7258064516129032\n",
            "332.0 434 0.7649769585253456\n",
            "312.0 432 0.7222222222222222\n",
            "323.0 433 0.745958429561201\n",
            "321.0 433 0.7413394919168591\n",
            "[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0.6495238095238095, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0.6716141001855288, 0.5862708719851577, 0, 0, 0, 0, 0, 0, 0, 0], [0.518095238095238, 0.6393088552915767, 0.5027829313543599, 0, 0, 0, 0, 0, 0, 0], [0.6533333333333333, 0.7602591792656588, 0.6066790352504638, 0.6112311015118791, 0, 0, 0, 0, 0, 0], [0.64, 0.8056155507559395, 0.5658627087198516, 0.671264367816092, 0.7580993520518359, 0, 0, 0, 0, 0], [0.6171428571428571, 0.7688984881209503, 0.5788497217068646, 0.6451612903225806, 0.7494600431965442, 0.8551724137931035, 0, 0, 0, 0], [0.6019047619047619, 0.7473002159827213, 0.5677179962894249, 0.7860576923076923, 0.7386609071274298, 0.8, 0.7857142857142857, 0, 0, 0], [0.6228571428571429, 0.7170626349892009, 0.5844155844155844, 0.6273148148148148, 0.7084233261339092, 0.7586206896551724, 0.7258064516129032, 0.7222222222222222, 0, 0], [0.5961904761904762, 0.7775377969762419, 0.5398886827458256, 0.6674364896073903, 0.7278617710583153, 0.7816091954022989, 0.7649769585253456, 0.745958429561201, 0.7413394919168591, 0]]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xGrZ2AVMB-5y",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "c668b8ae-2a0d-41a7-85c6-f25f35003e8c"
      },
      "source": [
        "#q1b.py - attempt 1\n",
        "import csv\n",
        "import os\n",
        "from Bio import SeqIO\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from Bio.Seq import Seq\n",
        "from Bio import pairwise2\n",
        "\n",
        "def merge(list1, list2):\n",
        "      \n",
        "    merged_list = [(list1[i], list2[i]) for i in range(0, len(list1))]\n",
        "    return merged_list\n",
        "\n",
        "# Python program to get transpose\n",
        "# elements of two dimension list\n",
        "def transpose(l1, l2):\n",
        " \n",
        "    # iterate over list l1 to the length of an item\n",
        "    for i in range(len(l1[0])):\n",
        "        # print(i)\n",
        "        row =[]\n",
        "        for item in l1:\n",
        "            # appending to new list with values and index positions\n",
        "            # i contains index position and item contains values\n",
        "            row.append(item[i])\n",
        "        l2.append(row)\n",
        "    return l2\n",
        "\n",
        "def main ():\n",
        "#to help resolve indent errors \n",
        "    file_in ='./Ndistance.txt'\n",
        "    file_out=''\n",
        "\n",
        "    if os.access(file_in, os.R_OK):\n",
        "        print (\"File \", file_in ,\" is accessible to read\")\n",
        "    else:\n",
        "        print (\"File \", file_in ,\" is not accessible to read, exiting.\")\n",
        "        exit (1)\n",
        "    reader = csv.reader(open(file_in,\"r+\"), delimiter=',')\n",
        "    distance = list(reader)\n",
        "\n",
        "    print(distance)\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "\t  main()"
      ],
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "File  ./Ndistance.txt  is accessible to read\n",
            "[['0', '0', '0', '0', '0', '0', '0', '0', '0', '0'], ['0.6495238095238095', '0', '0', '0', '0', '0', '0', '0', '0', '0'], ['0.6716141001855288', '0.5862708719851577', '0', '0', '0', '0', '0', '0', '0', '0'], ['0.518095238095238', '0.6393088552915767', '0.5027829313543599', '0', '0', '0', '0', '0', '0', '0'], ['0.6533333333333333', '0.7602591792656588', '0.6066790352504638', '0.6112311015118791', '0', '0', '0', '0', '0', '0'], ['0.64', '0.8056155507559395', '0.5658627087198516', '0.671264367816092', '0.7580993520518359', '0', '0', '0', '0', '0'], ['0.6171428571428571', '0.7688984881209503', '0.5788497217068646', '0.6451612903225806', '0.7494600431965442', '0.8551724137931035', '0', '0', '0', '0'], ['0.6019047619047619', '0.7473002159827213', '0.5677179962894249', '0.7860576923076923', '0.7386609071274298', '0.8', '0.7857142857142857', '0', '0', '0'], ['0.6228571428571429', '0.7170626349892009', '0.5844155844155844', '0.6273148148148148', '0.7084233261339092', '0.7586206896551724', '0.7258064516129032', '0.7222222222222222', '0', '0'], ['0.5961904761904762', '0.7775377969762419', '0.5398886827458256', '0.6674364896073903', '0.7278617710583153', '0.7816091954022989', '0.7649769585253456', '0.745958429561201', '0.7413394919168591', '0']]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "BtN14O-0w8vn",
        "outputId": "e959c74f-64b5-47e7-b885-c4fd9f330ff6"
      },
      "source": [
        "#q1b.py - attempt 2\n",
        "\"\"\"\n",
        "UPGMA - Unweighted pair group method using arithmetic averages\n",
        "Takes as input the distance matrix of species as a numpy array\n",
        "Returns tree either as a dendrogram or in Newick format\n",
        "\n",
        "Resources reffered to and used :\n",
        "\n",
        "1. https://stackoverflow.com/questions/58532758/read-csv-file-into-list-and-convert-the-strings-into-integers-python \n",
        "2. http://etetoolkit.org/treeview/ \n",
        "3. https://www.geeksforgeeks.org/writing-data-from-a-python-list-to-csv-row-wise/ \n",
        "4. https://stackoverflow.com/questions/56025046/2d-list-to-csv-by-column \n",
        "5. https://en.wikipedia.org/wiki/UPGMA \n",
        "6. https://en.wikipedia.org/wiki/Phylogenetic_tree \n",
        "7. https://en.wikipedia.org/wiki/Child_node \n",
        "8. https://en.wikipedia.org/wiki/Tree_structure \n",
        "9. https://medium.com/@sharma.ravit/upgma-method-designing-a-phylogenetic-tree-9a708de18419#:~:text=%20UPGMA%20Method%3A%20Designing%20a%20Phylogenetic%20Tree%20,1%E2%80%932%20until%20the%20tree%20is%20complete%20More%20 \n",
        "10. https://bit.ly/39T1tmb \n",
        "11. https://telliott99.blogspot.com/2010/11/upgma-in-python.html \n",
        "12. https://telliott99.blogspot.com/2010/11/upgma-in-python-2.html \n",
        "13. https://telliott99.blogspot.com/2010/11/upgma-in-python-3-sarich-data.html \n",
        "14. https://telliott99.blogspot.com/2010/03/clustering-with-upgma.html \n",
        "15. https://telliott99.blogspot.com/2010/04/visualizing-upgma-clustering.html \n",
        "16. https://www.geeksforgeeks.org/draw-a-tree-using-arcade-library-in-python/ \n",
        "17. https://www.geeksforgeeks.org/binarytree-module-in-python/ \n",
        "18. https://github.com/itsjinendrajain/Coding-Ninjas-Problem-Solving-Using-Python/tree/main/8.Searching%20%26%20Sorting \n",
        "19. https://www.geeksforgeeks.org/tree-sort/ \n",
        "20. http://www.bx.psu.edu/~dcking/man/newicktree.html#:~:text=The%20Newick%20Standard%20for%20representing%20trees%20in%20computer-readable,represented%20by%20the%20following%20sequence%20of%20printable%20characters%3A \n",
        "21. https://stackoverflow.com/questions/61117131/how-to-convert-a-binary-tree-to-a-newick-tree-using-python \n",
        "22. https://pure.mpg.de/rest/items/item_3258810_1/component/file_3273695/content \n",
        "23. http://etetoolkit.org/docs/2.3/tutorial/tutorial_trees.html \n",
        "\n",
        "\"\"\"\n",
        "\n",
        "#%% import modules\n",
        "import numpy as np\n",
        "import itertools\n",
        "import csv\n",
        "import os\n",
        "from Bio import SeqIO\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from Bio.Seq import Seq\n",
        "from Bio import pairwise2\n",
        "import pandas as pd\n",
        "\n",
        "def merge(list1, list2):\n",
        "      \n",
        "    merged_list = [(list1[i], list2[i]) for i in range(0, len(list1))]\n",
        "    return merged_list\n",
        "\n",
        "# Python program to get transpose\n",
        "# elements of two dimension list\n",
        "def transpose(l1, l2):\n",
        " \n",
        "    # iterate over list l1 to the length of an item\n",
        "    for i in range(len(l1[0])):\n",
        "        # print(i)\n",
        "        row =[]\n",
        "        for item in l1:\n",
        "            # appending to new list with values and index positions\n",
        "            # i contains index position and item contains values\n",
        "            row.append(item[i])\n",
        "        l2.append(row)\n",
        "    return l2\n",
        "\n",
        "# class for nodes\n",
        "class Node:\n",
        "    \"\"\"\n",
        "    Data structure to store node of a UPGMA tree\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self, left=None, right=None, up_height=0.0, down_height=0.0):\n",
        "        \"\"\"\n",
        "        Creating a node.\n",
        "        For a single taxon, set taxon name as self.left, leave right as none.\n",
        "        For an operational taxonomic unit(OTU) set left and right to child nodes.\n",
        "\n",
        "        Parameters\n",
        "        ----------\n",
        "        left : default = none, taxon label\n",
        "        right : default = none, taxon label\n",
        "        up_height : float, default = 0.0, dist to parent node, if any\n",
        "        down_height : float, default = 0.0, dist to child node, if any\n",
        "        \"\"\"\n",
        "        self.left = left\n",
        "        self.right = right\n",
        "        self.uh = up_height\n",
        "        self.dh = down_height\n",
        "\n",
        "    def leaves(self) -> list:\n",
        "        \"\"\"\n",
        "        Method to find the taxa under any given node, effectively equivalent to\n",
        "        finding leaves of a binary tree. Only lists original taxa and not OTUs.\n",
        "\n",
        "        Returns a list of node names, not nodes themselves.\n",
        "        \"\"\"\n",
        "        if self == None:\n",
        "            return []\n",
        "        if self.right == None:\n",
        "            return [self.left]\n",
        "        leaves = self.left.leaves() + self.right.leaves()\n",
        "        return leaves\n",
        "\n",
        "    def __len__(self) -> int:\n",
        "        \"\"\"\n",
        "        Method to define len() of a node.\n",
        "\n",
        "        Returns the number of original taxa under any given node.\n",
        "        \"\"\"\n",
        "        return sum(1 for taxa in self.leaves())\n",
        "\n",
        "    def __repr__(self) -> str:\n",
        "        \"\"\"\n",
        "        Method to give readable print output\n",
        "        \"\"\"\n",
        "        return \"-\".join(self.leaves())\n",
        "\n",
        "\n",
        "# class for UPGMA\n",
        "class UPGMA:\n",
        "    def __init__(self, dist_matrix: np.ndarray, taxa: list):\n",
        "        \"\"\"\n",
        "        Initialize an UPGMA class.\n",
        "        Takes a nxn distance matrix as input. A list of n taxon id is required\n",
        "        in the same order as the distance matrix row/column\n",
        "\n",
        "        Parameters\n",
        "        ----------\n",
        "        dist_matrix : numpy array, distance matrix of species\n",
        "        taxa : list of int or str to identify taxa\n",
        "        \"\"\"\n",
        "        self.distances = dist_matrix\n",
        "        self.taxa = taxa\n",
        "        self.build_tree(self.distances, self.taxa)\n",
        "\n",
        "    def build_tree(self, dist_matrix: np.ndarray, taxa: list) -> Node:\n",
        "        \"\"\"\n",
        "        Method to construct a tree from a given distance matrix and taxa list.\n",
        "\n",
        "        Parameters\n",
        "        ----------\n",
        "        dist_matrix : np.ndarray of pairwise distances\n",
        "        taxa : list of taxa id. Elements of lists have to be unique\n",
        "\n",
        "        Returns the root node for constructed tree.\n",
        "        \"\"\"\n",
        "        # creating node for each taxa\n",
        "        nodes = list(Node(taxon) for taxon in taxa)\n",
        "\n",
        "        # dictionary row/column id -> node\n",
        "        rc_to_node = dict([i, j] for i, j in enumerate(nodes))\n",
        "        \n",
        "        # dictionary taxa -> row/column id\n",
        "        taxa_to_rc = dict([i, j] for j, i in enumerate(taxa))\n",
        "        \n",
        "        # make copy of dist matrix to work on\n",
        "        work_matrix = dist_matrix\n",
        "        # set all diagonal elements to infinity for ease of finding least distance\n",
        "        work_matrix = np.array(work_matrix, dtype=float)\n",
        "        np.fill_diagonal(work_matrix, np.inf)\n",
        "\n",
        "        # loop\n",
        "        while len(nodes) > 1:\n",
        "            # finding (row, col) of least dist\n",
        "            least_id = np.unravel_index(work_matrix.argmin(), work_matrix.shape, \"C\")\n",
        "            least_dist = work_matrix[least_id[0], least_id[1]]\n",
        "            # nodes corresponding to (row, col)\n",
        "            node1, node2 = rc_to_node[least_id[0]], rc_to_node[least_id[1]]\n",
        "            \n",
        "            # add OTU with node1 and node2 as children. set heights of nodes\n",
        "            new_node = Node(node2, node1)\n",
        "            nodes.append(new_node)\n",
        "            node1.uh = least_dist / 2 - node1.dh\n",
        "            node2.uh = least_dist / 2 - node2.dh\n",
        "            new_node.dh = least_dist / 2\n",
        "            nodes.remove(node1)\n",
        "            nodes.remove(node2)\n",
        "           \n",
        "            # create new working distance matrix\n",
        "            work_matrix = self.update_distance(dist_matrix, nodes, taxa_to_rc)\n",
        "            \n",
        "            # update row/col id -> node dictionary\n",
        "            rc_to_node = dict([i, j] for i, j in enumerate(nodes))\n",
        "        # set tree to root\n",
        "        self.tree = nodes[0]\n",
        "\n",
        "    def update_distance(\n",
        "        self, dist_matrix: np.ndarray, nodes: list, taxa_to_rc: dict\n",
        "    ) -> np.ndarray:\n",
        "        \"\"\"\n",
        "        Method to make a new distance matrix with newer node list.\n",
        "\n",
        "        Parameters\n",
        "        ----------\n",
        "        dist_matrix : np.ndarray of pairwise distances for all taxa\n",
        "        nodes : list of updated nodes\n",
        "        taxa_to_rc : dict for taxa -> row/col id\n",
        "\n",
        "        Returns np.ndarray of pairwise distances for updated nodes\n",
        "        \"\"\"\n",
        "        # dictionary for node -> row/col id\n",
        "        node_to_rc = dict([i, j] for j, i in enumerate(nodes))\n",
        "        \n",
        "        rc = len(nodes)\n",
        "        new_dist_matrix = np.zeros((rc, rc), dtype=float)\n",
        "        for node1 in nodes:\n",
        "            row = node_to_rc[node1]\n",
        "            for node2 in nodes:\n",
        "                node_pairs = list(itertools.product(node1.leaves(), node2.leaves()))\n",
        "                col = node_to_rc[node2]\n",
        "                new_dist_matrix[row, col] = sum(\n",
        "                    dist_matrix[taxa_to_rc[i], taxa_to_rc[j]] for i, j in node_pairs\n",
        "                ) / len(node_pairs)\n",
        "        np.fill_diagonal(new_dist_matrix, np.inf)\n",
        "        return new_dist_matrix\n",
        "\n",
        "\n",
        "def tree_to_newick(t) -> str:\n",
        "    \"\"\"\n",
        "    Function to convert tree to Newick, slightly modified form of the tree.py version.\n",
        "    Takes the root node of an UPGMA tree as input\n",
        "    \"\"\"\n",
        "    if t.right == None:\n",
        "        return t.left + \":\" + str(t.uh)\n",
        "    else:\n",
        "        return (\n",
        "            \"(\"\n",
        "            + \",\".join([tree_to_newick(x) for x in [t.left, t.right]])\n",
        "            + \"):\"\n",
        "            + str(t.uh)\n",
        "        )\n",
        "\n",
        "\n",
        "#main\n",
        "\n",
        "def main ():\n",
        "#to help resolve indent errors \n",
        "    file_in ='./Ndistance.txt'\n",
        "    file_in_2='./Ntaxa.txt'\n",
        "    file_out=\"./Nnewick.txt\"\n",
        "    if os.access(file_in, os.R_OK):\n",
        "        print (\"File \", file_in ,\" is accessible to read\")\n",
        "    else:\n",
        "        print (\"File \", file_in ,\" is not accessible to read, exiting.\")\n",
        "        exit (1)\n",
        "    reader = csv.reader(open(file_in,\"r+\"), delimiter=',')\n",
        "    distance = list(reader)\n",
        "    print(distance)\n",
        "    outer_out_list = []\n",
        "    for inner_list in distance:\n",
        "        innet_out_list = []\n",
        "        for string in inner_list:\n",
        "            innet_out_list.append(float (string))\n",
        "        outer_out_list.append(innet_out_list)\n",
        "    print(outer_out_list)\n",
        "    \n",
        "    if os.access(file_in_2, os.R_OK):\n",
        "        print (\"File \", file_in_2 ,\" is accessible to read\\n\")\n",
        "    else:\n",
        "        print (\"File \", file_in_2 ,\" is not accessible to read, exiting.\")\n",
        "        exit (1)    \n",
        "    reader_2 = csv.reader(open(file_in_2,\"r+\"), delimiter=',')\n",
        "    taxa = list(reader_2)\n",
        "    taxa = taxa[0]\n",
        "    print( taxa )\n",
        "    \n",
        "    distances = np.array( outer_out_list )\n",
        "    x = UPGMA(distances, taxa).tree\n",
        "    final_output_newick=(tree_to_newick(x))\n",
        "    print(final_output_newick)\n",
        "    with open(file_out,\"w\") as f_out:\n",
        "      f_out.write(final_output_newick)\n",
        "    \n",
        "if __name__ == \"__main__\":\n",
        "    main()"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "File  ./Ndistance.txt  is accessible to read\n",
            "[['0', '0', '0', '0', '0', '0', '0', '0', '0', '0'], ['0.6495238095238095', '0', '0', '0', '0', '0', '0', '0', '0', '0'], ['0.6716141001855288', '0.5862708719851577', '0', '0', '0', '0', '0', '0', '0', '0'], ['0.518095238095238', '0.6393088552915767', '0.5027829313543599', '0', '0', '0', '0', '0', '0', '0'], ['0.6533333333333333', '0.7602591792656588', '0.6066790352504638', '0.6112311015118791', '0', '0', '0', '0', '0', '0'], ['0.64', '0.8056155507559395', '0.5658627087198516', '0.671264367816092', '0.7580993520518359', '0', '0', '0', '0', '0'], ['0.6171428571428571', '0.7688984881209503', '0.5788497217068646', '0.6451612903225806', '0.7494600431965442', '0.8551724137931035', '0', '0', '0', '0'], ['0.6019047619047619', '0.7473002159827213', '0.5677179962894249', '0.7860576923076923', '0.7386609071274298', '0.8', '0.7857142857142857', '0', '0', '0'], ['0.6228571428571429', '0.7170626349892009', '0.5844155844155844', '0.6273148148148148', '0.7084233261339092', '0.7586206896551724', '0.7258064516129032', '0.7222222222222222', '0', '0'], ['0.5961904761904762', '0.7775377969762419', '0.5398886827458256', '0.6674364896073903', '0.7278617710583153', '0.7816091954022989', '0.7649769585253456', '0.745958429561201', '0.7413394919168591', '0']]\n",
            "[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.6495238095238095, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.6716141001855288, 0.5862708719851577, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.518095238095238, 0.6393088552915767, 0.5027829313543599, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.6533333333333333, 0.7602591792656588, 0.6066790352504638, 0.6112311015118791, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [0.64, 0.8056155507559395, 0.5658627087198516, 0.671264367816092, 0.7580993520518359, 0.0, 0.0, 0.0, 0.0, 0.0], [0.6171428571428571, 0.7688984881209503, 0.5788497217068646, 0.6451612903225806, 0.7494600431965442, 0.8551724137931035, 0.0, 0.0, 0.0, 0.0], [0.6019047619047619, 0.7473002159827213, 0.5677179962894249, 0.7860576923076923, 0.7386609071274298, 0.8, 0.7857142857142857, 0.0, 0.0, 0.0], [0.6228571428571429, 0.7170626349892009, 0.5844155844155844, 0.6273148148148148, 0.7084233261339092, 0.7586206896551724, 0.7258064516129032, 0.7222222222222222, 0.0, 0.0], [0.5961904761904762, 0.7775377969762419, 0.5398886827458256, 0.6674364896073903, 0.7278617710583153, 0.7816091954022989, 0.7649769585253456, 0.745958429561201, 0.7413394919168591, 0.0]]\n",
            "File  ./Ntaxa.txt  is accessible to read\n",
            "\n",
            "['NM_205222.3', 'NM_001130093.2', 'KX951416.1', 'BT006808.1', 'NM_019129.3', 'NM_001109772.1', 'NM_173926.2', 'NM_001008996.2', 'M57671.1', 'U03610.1']\n",
            "(((NM_001008996.2:0.0,NM_173926.2:0.0):0.0,(NM_001109772.1:0.0,NM_019129.3:0.0):0.0):0.12365581342023683,((U03610.1:0.0,M57671.1:0.0):0.0,((BT006808.1:0.0,KX951416.1:0.0):0.0,(NM_001130093.2:0.0,NM_205222.3:0.0):0.0):0.0):0.12365581342023683):0.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XLQWEZhp8HPB"
      },
      "source": [
        "# Now, we have completed question 1!!!\n",
        "\n",
        "# Time for question 2 !! "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 243
        },
        "id": "pYoSW0W1yRTX",
        "outputId": "40998f2b-bd7f-4bb2-f874-82b42a68945f"
      },
      "source": [
        "#Create Protien.txt ,Pdistance.txt and BLOSUM62 to run q2a.py and q2b.py\n",
        "\n",
        "\n",
        "\n",
        "#Copy paste whatever Protien.txt should contain, use FASTA format.\n",
        "#Instead of running this code block you may also upload the files into the $PWD of google colab.\n",
        "\n",
        "ProteinTextInput='''>NP_990553.1 insulin preproinsulin precursor [Gallus gallus]\n",
        "MALWIRSLPLLALLVFSGPGTSYAAANQHLCGSHLVEALYLVCGERGFFYSPKARRDVEQPLVSSPLRGE\n",
        "AGVLPFQQEEYEKVKRGIVEQCCHNTCSLYQLENYCN\n",
        "\n",
        ">NP_001123565.1 insulin precursor [Canis lupus familiaris]\n",
        "MALWMRLLPLLALLALWAPAPTRAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREVEDLQVRDVELAG\n",
        "APGEGGLQPLALEGALQKRGIVEQCCTSICSLYQLENYCN\n",
        "\n",
        ">ATN38177.1 insulin [Labeo rohita]\n",
        "MAVWLQAGALLFLLAVSSVNANAGAPQHLCGSHLVDALYLVCGPTGFFYNPKRDVDPLMGFLPPKSAQET\n",
        "EVADFAFKDHAEVIRKRGIVEQCCHKPCSIFELQNYCN\n",
        "\n",
        ">AAP35454.1 insulin [Homo sapiens]\n",
        "MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGG\n",
        "GPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
        "\n",
        ">NP_062002.1 insulin-1 preproprotein [Rattus norvegicus]\n",
        "MALWMRFLPLLALLVLWEPKPAQAFVKQHLCGPHLVEALYLVCGERGFFYTPKSRREVEDPQVPQLELGG\n",
        "GPEAGDLQTLALEVARQKRGIVDQCCTSICSLYQLENYCN\n",
        "\n",
        ">NP_001103242.1 insulin precursor [Sus scrofa]\n",
        "MALWTRLLPLLALLALWAPAPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREAENPQAGAVELGG\n",
        "GLGGLQALALEGPPQKRGIVEQCCTSICSLYQLENYCN\n",
        "\n",
        ">NP_776351.2 insulin preproprotein [Bos taurus]\n",
        "MALWTRLAPLLALLALWAPAPARAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREVEGPQVGALELAG\n",
        "GPGAGGLEGPPQKRGIVEQCCASVCSLYQLENYCN\n",
        "\n",
        ">NP_001008996.1 insulin preproprotein [Pan troglodytes]\n",
        "MALWMRLLPLLVLLALWGPDPASAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGG\n",
        "GPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
        "\n",
        ">AAA40590.1 insulin [Octodon degus]\n",
        "MAPWMHLLTVLALLALWGPNSVQAYSSQHLCGSNLVEALYMTCGRSGFYRPHDRRELEDLQVEQAELGLE\n",
        "AGGLQPSALEMILQKRGIVDQCCNNICTFNQLQNYCNVP\n",
        "\n",
        ">AAA19033.1 insulin [Oryctolagus cuniculus]\n",
        "MASLAALLPLLALLVLCRLDPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKSRREVEELQVGQAELGG\n",
        "GPGAGGLQPSALELALQKRGIVEQCCTSICSLYQLENYCN\n",
        "\n",
        "'''\n",
        "\n",
        "#Copy paste whatever Blosum62.txt should contain, remove the header as its irrelevant.\n",
        "#Instead of running this code block you may also upload the files into the $PWD of google colab.\n",
        "\n",
        "Blosum62TextInput='''*  A  R  N  D  C  Q  E  G  H  I  L  K  M  F  P  S  T  W  Y  V  B  Z  X  *\n",
        "A  4 -1 -2 -2  0 -1 -1  0 -2 -1 -1 -1 -1 -2 -1  1  0 -3 -2  0 -2 -1  0 -4 \n",
        "R -1  5  0 -2 -3  1  0 -2  0 -3 -2  2 -1 -3 -2 -1 -1 -3 -2 -3 -1  0 -1 -4 \n",
        "N -2  0  6  1 -3  0  0  0  1 -3 -3  0 -2 -3 -2  1  0 -4 -2 -3  3  0 -1 -4 \n",
        "D -2 -2  1  6 -3  0  2 -1 -1 -3 -4 -1 -3 -3 -1  0 -1 -4 -3 -3  4  1 -1 -4 \n",
        "C  0 -3 -3 -3  9 -3 -4 -3 -3 -1 -1 -3 -1 -2 -3 -1 -1 -2 -2 -1 -3 -3 -2 -4 \n",
        "Q -1  1  0  0 -3  5  2 -2  0 -3 -2  1  0 -3 -1  0 -1 -2 -1 -2  0  3 -1 -4 \n",
        "E -1  0  0  2 -4  2  5 -2  0 -3 -3  1 -2 -3 -1  0 -1 -3 -2 -2  1  4 -1 -4 \n",
        "G  0 -2  0 -1 -3 -2 -2  6 -2 -4 -4 -2 -3 -3 -2  0 -2 -2 -3 -3 -1 -2 -1 -4 \n",
        "H -2  0  1 -1 -3  0  0 -2  8 -3 -3 -1 -2 -1 -2 -1 -2 -2  2 -3  0  0 -1 -4 \n",
        "I -1 -3 -3 -3 -1 -3 -3 -4 -3  4  2 -3  1  0 -3 -2 -1 -3 -1  3 -3 -3 -1 -4 \n",
        "L -1 -2 -3 -4 -1 -2 -3 -4 -3  2  4 -2  2  0 -3 -2 -1 -2 -1  1 -4 -3 -1 -4 \n",
        "K -1  2  0 -1 -3  1  1 -2 -1 -3 -2  5 -1 -3 -1  0 -1 -3 -2 -2  0  1 -1 -4 \n",
        "M -1 -1 -2 -3 -1  0 -2 -3 -2  1  2 -1  5  0 -2 -1 -1 -1 -1  1 -3 -1 -1 -4 \n",
        "F -2 -3 -3 -3 -2 -3 -3 -3 -1  0  0 -3  0  6 -4 -2 -2  1  3 -1 -3 -3 -1 -4 \n",
        "P -1 -2 -2 -1 -3 -1 -1 -2 -2 -3 -3 -1 -2 -4  7 -1 -1 -4 -3 -2 -2 -1 -2 -4 \n",
        "S  1 -1  1  0 -1  0  0  0 -1 -2 -2  0 -1 -2 -1  4  1 -3 -2 -2  0  0  0 -4 \n",
        "T  0 -1  0 -1 -1 -1 -1 -2 -2 -1 -1 -1 -1 -2 -1  1  5 -2 -2  0 -1 -1  0 -4 \n",
        "W -3 -3 -4 -4 -2 -2 -3 -2 -2 -3 -2 -3 -1  1 -4 -3 -2 11  2 -3 -4 -3 -2 -4 \n",
        "Y -2 -2 -2 -3 -2 -1 -2 -3  2 -1 -1 -2 -1  3 -3 -2 -2  2  7 -1 -3 -2 -1 -4 \n",
        "V  0 -3 -3 -3 -1 -2 -2 -3 -3  3  1 -2  1 -1 -2 -2  0 -3 -1  4 -3 -2 -1 -4 \n",
        "B -2 -1  3  4 -3  0  1 -1  0 -3 -4  0 -3 -3 -2  0 -1 -4 -3 -3  4  1 -1 -4 \n",
        "Z -1  0  0  1 -3  3  4 -2  0 -3 -3  1 -1 -3 -1  0 -1 -3 -2 -2  1  4 -1 -4 \n",
        "X  0 -1 -1 -1 -2 -1 -1 -1 -1 -1 -1 -1 -1 -1 -2  0  0 -2 -1 -1 -1 -1 -1 -4 \n",
        "* -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4 -4  1'''\n",
        "\n",
        "\n",
        "\n",
        "with open(\"Protein.txt\", 'x') as f_out:\n",
        "  f_out.write( ProteinTextInput )\n",
        "\n",
        "with open(\"Pdistance.txt\",'x') as f_out_2:\n",
        "  f_out_2.write('')\n",
        "\n",
        "with open(\"BLOSUM62.txt\",'x') as f_out_3:\n",
        "  f_out_3.write(Blosum62TextInput)\n",
        "\n",
        "\n",
        "exit(0)\n"
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "error",
          "ename": "FileExistsError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFileExistsError\u001b[0m                           Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-1-b8eaef3c9ab6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     79\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     80\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 81\u001b[0;31m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Protein.txt\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'x'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf_out\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     82\u001b[0m   \u001b[0mf_out\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0mProteinTextInput\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     83\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFileExistsError\u001b[0m: [Errno 17] File exists: 'Protein.txt'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lWxFdzZ8ACJ9",
        "outputId": "8c2a9f8e-02e1-4853-ce5e-7663c4ff0f26"
      },
      "source": [
        "#q2a.py\n",
        "'''\n",
        " After reading given input of Nucleotide.txt, I see that each new entry begins with a '>',\n",
        " followed by a line of Information on the gene, followed by some N lines of ATGC\n",
        "\n",
        " Identified as the FASTA format.\n",
        "\n",
        " I shall try to use biopython to give us the (n^2 - n)/2 values for distance matrix.\n",
        "\n",
        " Resources referred to \n",
        " \n",
        " 1. https://python.omics.wiki/biopython/examples/read-fasta\n",
        " 2. https://www.geeksforgeeks.org/biopython-pairwise-alignment/\n",
        " 3. https://stackoverflow.com/questions/20580657/how-to-read-a-fasta-file-in-python \n",
        " 4. https://vlab.amrita.edu/index.php?sub=3&brch=274&sim=1438&cnt=1 \n",
        " 5. https://www.ebi.ac.uk/seqdb/confluence/display/JDSAT/Clustal+Omega+Help+and+Documentation#ClustalOmegaHelpandDocumentation-sequence \n",
        " 6. https://www.tutorialspoint.com/biopython/biopython_sequence_alignments.htm \n",
        " 7. https://www.geeksforgeeks.org/biopython-sequence-alignment/#:~:text=Biopython%20%E2%80%93%20Sequence%20Alignment%20Last%20Updated%20%3A%2011,identify%20the%20region%20of%20similarity%20among%20them.%20 \n",
        " 8. https://www.ncbi.nlm.nih.gov/sra/docs/submitformats/#:~:text=fasta%20files%20may%20be%20submitted%20with%20corresponding%20qual%20files%2C%20too.%20these%20are%20recognized%20in%20the%20sra%20data%20processing%20pipeline%20as%20equivalent%20to%20fastq%20and%20should%20be%20specified%20as%20fastq%20when%20submitting%20the%20data%20files. \n",
        " 9. https://www.csestack.org/python-check-if-file-directory-exists/\n",
        " 10. https://flexiple.com/python-create-file/ \n",
        " 11. https://www.askpython.com/python/built-in-methods/python-print-to-file#:~:text=Python%20%E2%80%93%20Print%20to%20File%201%20Method%201%3A,4%20Method%204%3A%20Use%20the%20logging%20module.%20 \n",
        " 12. https://stackoverflow.com/questions/9426045/difference-between-exit0-and-exit1-in-python#:~:text=exit%20%281%29%20means%20there%20was%20some%20issue%20%2F,A%20zero%20error%20code%20means%20a%20successful%20exit. \n",
        "\n",
        "\n",
        "'''\n",
        "\n",
        "import csv\n",
        "import os\n",
        "from Bio import SeqIO\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from Bio.Seq import Seq\n",
        "from Bio import pairwise2\n",
        "import argparse\n",
        "import pathlib\n",
        "from Bio.PDB import FastMMCIFParser, MMCIFIO, PDBParser, PDBIO, Superimposer\n",
        "from Bio.PDB.Polypeptide import is_aa\n",
        "from Bio.Align import substitution_matrices\n",
        "from Bio.Data.SCOPData import protein_letters_3to1 as aa3to1\n",
        "\n",
        "def merge(list1, list2):\n",
        "      \n",
        "    merged_list = [(list1[i], list2[i]) for i in range(0, len(list1))]\n",
        "    return merged_list\n",
        "\n",
        "# Python program to get transpose\n",
        "# elements of two dimension list\n",
        "def transpose(l1, l2):\n",
        " \n",
        "    # iterate over list l1 to the length of an item\n",
        "    for i in range(len(l1[0])):\n",
        "        # print(i)\n",
        "        row =[]\n",
        "        for item in l1:\n",
        "            # appending to new list with values and index positions\n",
        "            # i contains index position and item contains values\n",
        "            row.append(item[i])\n",
        "        l2.append(row)\n",
        "    return l2\n",
        "\n",
        "def main ():\n",
        "#to help resolve indent errors \n",
        "    file_in ='./Protein.txt'\n",
        "    file_in_2='./BLOSUM62.txt'\n",
        "    file_out='./Pdistance.txt'\n",
        "\n",
        "    if os.access(file_in, os.R_OK):\n",
        "        print (\"File \", file_in ,\" is accessible to read\")\n",
        "    else:\n",
        "        print (\"File \", file_in ,\" is not accessible to read, exiting.\")\n",
        "        exit (1)\n",
        "\n",
        "    if os.access(file_in_2, os.R_OK):\n",
        "        print (\"File \", file_in_2 ,\" is accessible to read\")\n",
        "    else:\n",
        "        print (\"File \", file_in_2 ,\" is not accessible to read, exiting.\")\n",
        "        exit (1)\n",
        "    \n",
        "    if os.access(file_out, os.W_OK):\n",
        "        print (\"File \", file_out,\" is accessible to write \\n\")\n",
        "    else:\n",
        "        print (\"File \", file_out,\" is not accessible to write, exiting.\")\n",
        "        exit (1)\n",
        "\n",
        "    IdList = list ()\n",
        "    ProtienSeqList= list ()\n",
        "\n",
        "    with open(file_out, 'w') as f_out:\n",
        "    #to help resolve indent errors   \n",
        "        print(\"Beginning to Read \",file_in,\"\\n\")\n",
        "        for seq_record in SeqIO.parse(open(file_in, mode='r'), 'fasta'):\n",
        "        #to help resolve indent errors \n",
        "\n",
        "            # remove .id from .description record (remove all before first space)\n",
        "            seq_record.description=' '.join(seq_record.description.split()[1:])\n",
        "            \n",
        "            # do something (print or edit seq_record)       \n",
        "            print('SequenceID = '  + seq_record.id)\n",
        "            print('Description = ' + seq_record.description)\n",
        "            print('Sequence = ' + seq_record.seq + '\\n' )\n",
        "\n",
        "            # Save the Sequence into ProtienSeqList and Id into IdList\n",
        "            ProtienSeqList.append( seq_record.seq )\n",
        "            IdList.append( seq_record.id )\n",
        "\n",
        "        no_of_proteins = len(ProtienSeqList )\n",
        "        print(no_of_proteins,'Proteins scanned into our List\\n')\n",
        "        print(\"Our Lists are : \",merge ( IdList, ProtienSeqList ))\n",
        "        print(ProtienSeqList)\n",
        "        print(IdList,\"\\n\")\n",
        "\n",
        "        #Initialize distance matrix\n",
        "        dist_Matrix=list()\n",
        "\n",
        "        #Now to find distances, pairwise \n",
        "        for i in range(0,no_of_proteins,1):\n",
        "            column= list ()\n",
        "            for i_1 in range(0,i+1,1):\n",
        "                column.append(0)\n",
        "            for j in range (i+1, no_of_proteins, 1):\n",
        "                #compare with j'th protein, and find best pairing.\n",
        "                # alignments = pairwise2.align.globalxx(ProtienSeqList[i], ProtienSeqList[j])\n",
        "                test_alignments =  pairwise2.align.globalds( ProtienSeqList[i], ProtienSeqList[j] , substitution_matrices.load(\"BLOSUM62\"), one_alignment_only=True, open=-10.0, extend=-0.5, penalize_end_gaps=(False, False))\n",
        "                length = ( max(len(ProtienSeqList[i]), len(ProtienSeqList[j])) // 1)\n",
        "                score= 0\n",
        "                for alignment in test_alignments: \n",
        "                    # print( (alignment[2]//1) , (alignment[4]//1) , ( (alignment[2]//1)/ (alignment[4]//1) ) ) \n",
        "                    if ( (alignment[2]//1) >  score ):\n",
        "                        score = (alignment[2]//1)\n",
        "                norm_score = score/length\n",
        "                print(score,length,norm_score)\n",
        "                #add to 'column'\n",
        "                column.append(norm_score)\n",
        "            dist_Matrix.append(column)\n",
        "        out_Matrix=list ()\n",
        "        out_Matrix = transpose (dist_Matrix, out_Matrix)\n",
        "        print(out_Matrix)\n",
        "        writer = csv.writer(f_out)\n",
        "        # f_out.write(str (no_of_nucleotides)+\",\\n\")\n",
        "        writer.writerows(out_Matrix)\n",
        "    file_out_2=\"./Ptaxa.txt\"\n",
        "    with open(file_out_2, 'w') as f_out_2:\n",
        "      writer = csv.writer(f_out_2)\n",
        "      writer.writerow (IdList)\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    main()\n",
        "\n",
        "\n"
      ],
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "File  ./Protein.txt  is accessible to read\n",
            "File  ./BLOSUM62.txt  is accessible to read\n",
            "File  ./Pdistance.txt  is accessible to write \n",
            "\n",
            "Beginning to Read  ./Protein.txt \n",
            "\n",
            "SequenceID = NP_990553.1\n",
            "Description = insulin preproinsulin precursor [Gallus gallus]\n",
            "Sequence = MALWIRSLPLLALLVFSGPGTSYAAANQHLCGSHLVEALYLVCGERGFFYSPKARRDVEQPLVSSPLRGEAGVLPFQQEEYEKVKRGIVEQCCHNTCSLYQLENYCN\n",
            "\n",
            "SequenceID = NP_001123565.1\n",
            "Description = insulin precursor [Canis lupus familiaris]\n",
            "Sequence = MALWMRLLPLLALLALWAPAPTRAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREVEDLQVRDVELAGAPGEGGLQPLALEGALQKRGIVEQCCTSICSLYQLENYCN\n",
            "\n",
            "SequenceID = ATN38177.1\n",
            "Description = insulin [Labeo rohita]\n",
            "Sequence = MAVWLQAGALLFLLAVSSVNANAGAPQHLCGSHLVDALYLVCGPTGFFYNPKRDVDPLMGFLPPKSAQETEVADFAFKDHAEVIRKRGIVEQCCHKPCSIFELQNYCN\n",
            "\n",
            "SequenceID = AAP35454.1\n",
            "Description = insulin [Homo sapiens]\n",
            "Sequence = MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
            "\n",
            "SequenceID = NP_062002.1\n",
            "Description = insulin-1 preproprotein [Rattus norvegicus]\n",
            "Sequence = MALWMRFLPLLALLVLWEPKPAQAFVKQHLCGPHLVEALYLVCGERGFFYTPKSRREVEDPQVPQLELGGGPEAGDLQTLALEVARQKRGIVDQCCTSICSLYQLENYCN\n",
            "\n",
            "SequenceID = NP_001103242.1\n",
            "Description = insulin precursor [Sus scrofa]\n",
            "Sequence = MALWTRLLPLLALLALWAPAPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREAENPQAGAVELGGGLGGLQALALEGPPQKRGIVEQCCTSICSLYQLENYCN\n",
            "\n",
            "SequenceID = NP_776351.2\n",
            "Description = insulin preproprotein [Bos taurus]\n",
            "Sequence = MALWTRLAPLLALLALWAPAPARAFVNQHLCGSHLVEALYLVCGERGFFYTPKARREVEGPQVGALELAGGPGAGGLEGPPQKRGIVEQCCASVCSLYQLENYCN\n",
            "\n",
            "SequenceID = NP_001008996.1\n",
            "Description = insulin preproprotein [Pan troglodytes]\n",
            "Sequence = MALWMRLLPLLVLLALWGPDPASAFVNQHLCGSHLVEALYLVCGERGFFYTPKTRREAEDLQVGQVELGGGPGAGSLQPLALEGSLQKRGIVEQCCTSICSLYQLENYCN\n",
            "\n",
            "SequenceID = AAA40590.1\n",
            "Description = insulin [Octodon degus]\n",
            "Sequence = MAPWMHLLTVLALLALWGPNSVQAYSSQHLCGSNLVEALYMTCGRSGFYRPHDRRELEDLQVEQAELGLEAGGLQPSALEMILQKRGIVDQCCNNICTFNQLQNYCNVP\n",
            "\n",
            "SequenceID = AAA19033.1\n",
            "Description = insulin [Oryctolagus cuniculus]\n",
            "Sequence = MASLAALLPLLALLVLCRLDPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKSRREVEELQVGQAELGGGPGAGGLQPSALELALQKRGIVEQCCTSICSLYQLENYCN\n",
            "\n",
            "10 Proteins scanned into our List\n",
            "\n",
            "Our Lists are :  [('NP_990553.1', Seq('MALWIRSLPLLALLVFSGPGTSYAAANQHLCGSHLVEALYLVCGERGFFYSPKA...YCN')), ('NP_001123565.1', Seq('MALWMRLLPLLALLALWAPAPTRAFVNQHLCGSHLVEALYLVCGERGFFYTPKA...YCN')), ('ATN38177.1', Seq('MAVWLQAGALLFLLAVSSVNANAGAPQHLCGSHLVDALYLVCGPTGFFYNPKRD...YCN')), ('AAP35454.1', Seq('MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKT...YCN')), ('NP_062002.1', Seq('MALWMRFLPLLALLVLWEPKPAQAFVKQHLCGPHLVEALYLVCGERGFFYTPKS...YCN')), ('NP_001103242.1', Seq('MALWTRLLPLLALLALWAPAPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKA...YCN')), ('NP_776351.2', Seq('MALWTRLAPLLALLALWAPAPARAFVNQHLCGSHLVEALYLVCGERGFFYTPKA...YCN')), ('NP_001008996.1', Seq('MALWMRLLPLLVLLALWGPDPASAFVNQHLCGSHLVEALYLVCGERGFFYTPKT...YCN')), ('AAA40590.1', Seq('MAPWMHLLTVLALLALWGPNSVQAYSSQHLCGSNLVEALYMTCGRSGFYRPHDR...NVP')), ('AAA19033.1', Seq('MASLAALLPLLALLVLCRLDPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKS...YCN'))]\n",
            "[Seq('MALWIRSLPLLALLVFSGPGTSYAAANQHLCGSHLVEALYLVCGERGFFYSPKA...YCN'), Seq('MALWMRLLPLLALLALWAPAPTRAFVNQHLCGSHLVEALYLVCGERGFFYTPKA...YCN'), Seq('MAVWLQAGALLFLLAVSSVNANAGAPQHLCGSHLVDALYLVCGPTGFFYNPKRD...YCN'), Seq('MALWMRLLPLLALLALWGPDPAAAFVNQHLCGSHLVEALYLVCGERGFFYTPKT...YCN'), Seq('MALWMRFLPLLALLVLWEPKPAQAFVKQHLCGPHLVEALYLVCGERGFFYTPKS...YCN'), Seq('MALWTRLLPLLALLALWAPAPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKA...YCN'), Seq('MALWTRLAPLLALLALWAPAPARAFVNQHLCGSHLVEALYLVCGERGFFYTPKA...YCN'), Seq('MALWMRLLPLLVLLALWGPDPASAFVNQHLCGSHLVEALYLVCGERGFFYTPKT...YCN'), Seq('MAPWMHLLTVLALLALWGPNSVQAYSSQHLCGSNLVEALYMTCGRSGFYRPHDR...NVP'), Seq('MASLAALLPLLALLVLCRLDPAQAFVNQHLCGSHLVEALYLVCGERGFFYTPKS...YCN')]\n",
            "['NP_990553.1', 'NP_001123565.1', 'ATN38177.1', 'AAP35454.1', 'NP_062002.1', 'NP_001103242.1', 'NP_776351.2', 'NP_001008996.1', 'AAA40590.1', 'AAA19033.1'] \n",
            "\n",
            "350.0 110 3.1818181818181817\n",
            "275.0 108 2.5462962962962963\n",
            "334.0 110 3.036363636363636\n",
            "342.0 110 3.109090909090909\n",
            "344.0 108 3.185185185185185\n",
            "337.0 107 3.149532710280374\n",
            "330.0 110 3.0\n",
            "254.0 109 2.330275229357798\n",
            "307.0 110 2.790909090909091\n",
            "242.0 110 2.2\n",
            "520.0 110 4.7272727272727275\n",
            "466.0 110 4.236363636363636\n",
            "481.0 110 4.372727272727273\n",
            "473.0 110 4.3\n",
            "516.0 110 4.6909090909090905\n",
            "347.0 110 3.1545454545454548\n",
            "453.0 110 4.118181818181818\n",
            "238.0 110 2.1636363636363636\n",
            "228.0 110 2.0727272727272728\n",
            "228.0 108 2.111111111111111\n",
            "244.0 108 2.259259259259259\n",
            "242.0 110 2.2\n",
            "204.0 109 1.871559633027523\n",
            "227.0 110 2.0636363636363635\n",
            "483.0 110 4.390909090909091\n",
            "484.0 110 4.4\n",
            "463.0 110 4.209090909090909\n",
            "582.0 110 5.290909090909091\n",
            "359.0 110 3.2636363636363637\n",
            "479.0 110 4.3545454545454545\n",
            "456.0 110 4.1454545454545455\n",
            "437.0 110 3.9727272727272727\n",
            "480.0 110 4.363636363636363\n",
            "339.0 110 3.081818181818182\n",
            "442.0 110 4.0181818181818185\n",
            "496.0 108 4.592592592592593\n",
            "481.0 110 4.372727272727273\n",
            "322.0 109 2.9541284403669725\n",
            "420.0 110 3.8181818181818183\n",
            "459.0 110 4.172727272727273\n",
            "291.0 109 2.669724770642202\n",
            "412.0 110 3.7454545454545456\n",
            "356.0 110 3.2363636363636363\n",
            "476.0 110 4.327272727272727\n",
            "332.0 110 3.018181818181818\n",
            "[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [3.1818181818181817, 0, 0, 0, 0, 0, 0, 0, 0, 0], [2.5462962962962963, 2.2, 0, 0, 0, 0, 0, 0, 0, 0], [3.036363636363636, 4.7272727272727275, 2.1636363636363636, 0, 0, 0, 0, 0, 0, 0], [3.109090909090909, 4.236363636363636, 2.0727272727272728, 4.390909090909091, 0, 0, 0, 0, 0, 0], [3.185185185185185, 4.372727272727273, 2.111111111111111, 4.4, 4.1454545454545455, 0, 0, 0, 0, 0], [3.149532710280374, 4.3, 2.259259259259259, 4.209090909090909, 3.9727272727272727, 4.592592592592593, 0, 0, 0, 0], [3.0, 4.6909090909090905, 2.2, 5.290909090909091, 4.363636363636363, 4.372727272727273, 4.172727272727273, 0, 0, 0], [2.330275229357798, 3.1545454545454548, 1.871559633027523, 3.2636363636363637, 3.081818181818182, 2.9541284403669725, 2.669724770642202, 3.2363636363636363, 0, 0], [2.790909090909091, 4.118181818181818, 2.0636363636363635, 4.3545454545454545, 4.0181818181818185, 3.8181818181818183, 3.7454545454545456, 4.327272727272727, 3.018181818181818, 0]]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "QpO44ZNqBjkE",
        "outputId": "59998479-dde7-42df-aada-99c6a8024659"
      },
      "source": [
        "#q2b.py - Basically a Copy-Paste of q1b.py\n",
        "\"\"\"\n",
        "UPGMA - Unweighted pair group method using arithmetic averages\n",
        "Takes as input the distance matrix of species as a numpy array\n",
        "Returns tree either as a dendrogram or in Newick format\n",
        "\n",
        "Resources reffered to and used :\n",
        "\n",
        "1. https://stackoverflow.com/questions/58532758/read-csv-file-into-list-and-convert-the-strings-into-integers-python \n",
        "2. http://etetoolkit.org/treeview/ \n",
        "3. https://www.geeksforgeeks.org/writing-data-from-a-python-list-to-csv-row-wise/ \n",
        "4. https://stackoverflow.com/questions/56025046/2d-list-to-csv-by-column \n",
        "5. https://en.wikipedia.org/wiki/UPGMA \n",
        "6. https://en.wikipedia.org/wiki/Phylogenetic_tree \n",
        "7. https://en.wikipedia.org/wiki/Child_node \n",
        "8. https://en.wikipedia.org/wiki/Tree_structure \n",
        "9. https://medium.com/@sharma.ravit/upgma-method-designing-a-phylogenetic-tree-9a708de18419#:~:text=%20UPGMA%20Method%3A%20Designing%20a%20Phylogenetic%20Tree%20,1%E2%80%932%20until%20the%20tree%20is%20complete%20More%20 \n",
        "10. https://bit.ly/39T1tmb \n",
        "11. https://telliott99.blogspot.com/2010/11/upgma-in-python.html \n",
        "12. https://telliott99.blogspot.com/2010/11/upgma-in-python-2.html \n",
        "13. https://telliott99.blogspot.com/2010/11/upgma-in-python-3-sarich-data.html \n",
        "14. https://telliott99.blogspot.com/2010/03/clustering-with-upgma.html \n",
        "15. https://telliott99.blogspot.com/2010/04/visualizing-upgma-clustering.html \n",
        "16. https://www.geeksforgeeks.org/draw-a-tree-using-arcade-library-in-python/ \n",
        "17. https://www.geeksforgeeks.org/binarytree-module-in-python/ \n",
        "18. https://github.com/itsjinendrajain/Coding-Ninjas-Problem-Solving-Using-Python/tree/main/8.Searching%20%26%20Sorting \n",
        "19. https://www.geeksforgeeks.org/tree-sort/ \n",
        "20. http://www.bx.psu.edu/~dcking/man/newicktree.html#:~:text=The%20Newick%20Standard%20for%20representing%20trees%20in%20computer-readable,represented%20by%20the%20following%20sequence%20of%20printable%20characters%3A \n",
        "21. https://stackoverflow.com/questions/61117131/how-to-convert-a-binary-tree-to-a-newick-tree-using-python \n",
        "22. https://pure.mpg.de/rest/items/item_3258810_1/component/file_3273695/content \n",
        "23. http://etetoolkit.org/docs/2.3/tutorial/tutorial_trees.html \n",
        "\n",
        "\"\"\"\n",
        "\n",
        "#%% import modules\n",
        "import numpy as np\n",
        "import itertools\n",
        "import csv\n",
        "import os\n",
        "from Bio import SeqIO\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from Bio.Seq import Seq\n",
        "from Bio import pairwise2\n",
        "import pandas as pd\n",
        "\n",
        "def merge(list1, list2):\n",
        "      \n",
        "    merged_list = [(list1[i], list2[i]) for i in range(0, len(list1))]\n",
        "    return merged_list\n",
        "\n",
        "# Python program to get transpose\n",
        "# elements of two dimension list\n",
        "def transpose(l1, l2):\n",
        " \n",
        "    # iterate over list l1 to the length of an item\n",
        "    for i in range(len(l1[0])):\n",
        "        # print(i)\n",
        "        row =[]\n",
        "        for item in l1:\n",
        "            # appending to new list with values and index positions\n",
        "            # i contains index position and item contains values\n",
        "            row.append(item[i])\n",
        "        l2.append(row)\n",
        "    return l2\n",
        "\n",
        "# class for nodes\n",
        "class Node:\n",
        "    \"\"\"\n",
        "    Data structure to store node of a UPGMA tree\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self, left=None, right=None, up_height=0.0, down_height=0.0):\n",
        "        \"\"\"\n",
        "        Creating a node.\n",
        "        For a single taxon, set taxon name as self.left, leave right as none.\n",
        "        For an operational taxonomic unit(OTU) set left and right to child nodes.\n",
        "\n",
        "        Parameters\n",
        "        ----------\n",
        "        left : default = none, taxon label\n",
        "        right : default = none, taxon label\n",
        "        up_height : float, default = 0.0, dist to parent node, if any\n",
        "        down_height : float, default = 0.0, dist to child node, if any\n",
        "        \"\"\"\n",
        "        self.left = left\n",
        "        self.right = right\n",
        "        self.uh = up_height\n",
        "        self.dh = down_height\n",
        "\n",
        "    def leaves(self) -> list:\n",
        "        \"\"\"\n",
        "        Method to find the taxa under any given node, effectively equivalent to\n",
        "        finding leaves of a binary tree. Only lists original taxa and not OTUs.\n",
        "\n",
        "        Returns a list of node names, not nodes themselves.\n",
        "        \"\"\"\n",
        "        if self == None:\n",
        "            return []\n",
        "        if self.right == None:\n",
        "            return [self.left]\n",
        "        leaves = self.left.leaves() + self.right.leaves()\n",
        "        return leaves\n",
        "\n",
        "    def __len__(self) -> int:\n",
        "        \"\"\"\n",
        "        Method to define len() of a node.\n",
        "\n",
        "        Returns the number of original taxa under any given node.\n",
        "        \"\"\"\n",
        "        return sum(1 for taxa in self.leaves())\n",
        "\n",
        "    def __repr__(self) -> str:\n",
        "        \"\"\"\n",
        "        Method to give readable print output\n",
        "        \"\"\"\n",
        "        return \"-\".join(self.leaves())\n",
        "\n",
        "\n",
        "# class for UPGMA\n",
        "class UPGMA:\n",
        "    def __init__(self, dist_matrix: np.ndarray, taxa: list):\n",
        "        \"\"\"\n",
        "        Initialize an UPGMA class.\n",
        "        Takes a nxn distance matrix as input. A list of n taxon id is required\n",
        "        in the same order as the distance matrix row/column\n",
        "\n",
        "        Parameters\n",
        "        ----------\n",
        "        dist_matrix : numpy array, distance matrix of species\n",
        "        taxa : list of int or str to identify taxa\n",
        "        \"\"\"\n",
        "        self.distances = dist_matrix\n",
        "        self.taxa = taxa\n",
        "        self.build_tree(self.distances, self.taxa)\n",
        "\n",
        "    def build_tree(self, dist_matrix: np.ndarray, taxa: list) -> Node:\n",
        "        \"\"\"\n",
        "        Method to construct a tree from a given distance matrix and taxa list.\n",
        "\n",
        "        Parameters\n",
        "        ----------\n",
        "        dist_matrix : np.ndarray of pairwise distances\n",
        "        taxa : list of taxa id. Elements of lists have to be unique\n",
        "\n",
        "        Returns the root node for constructed tree.\n",
        "        \"\"\"\n",
        "        # creating node for each taxa\n",
        "        nodes = list(Node(taxon) for taxon in taxa)\n",
        "\n",
        "        # dictionary row/column id -> node\n",
        "        rc_to_node = dict([i, j] for i, j in enumerate(nodes))\n",
        "        \n",
        "        # dictionary taxa -> row/column id\n",
        "        taxa_to_rc = dict([i, j] for j, i in enumerate(taxa))\n",
        "        \n",
        "        # make copy of dist matrix to work on\n",
        "        work_matrix = dist_matrix\n",
        "        # set all diagonal elements to infinity for ease of finding least distance\n",
        "        work_matrix = np.array(work_matrix, dtype=float)\n",
        "        np.fill_diagonal(work_matrix, np.inf)\n",
        "\n",
        "        # loop\n",
        "        while len(nodes) > 1:\n",
        "            # finding (row, col) of least dist\n",
        "            least_id = np.unravel_index(work_matrix.argmin(), work_matrix.shape, \"C\")\n",
        "            least_dist = work_matrix[least_id[0], least_id[1]]\n",
        "            # nodes corresponding to (row, col)\n",
        "            node1, node2 = rc_to_node[least_id[0]], rc_to_node[least_id[1]]\n",
        "            \n",
        "            # add OTU with node1 and node2 as children. set heights of nodes\n",
        "            new_node = Node(node2, node1)\n",
        "            nodes.append(new_node)\n",
        "            node1.uh = least_dist / 2 - node1.dh\n",
        "            node2.uh = least_dist / 2 - node2.dh\n",
        "            new_node.dh = least_dist / 2\n",
        "            nodes.remove(node1)\n",
        "            nodes.remove(node2)\n",
        "           \n",
        "            # create new working distance matrix\n",
        "            work_matrix = self.update_distance(dist_matrix, nodes, taxa_to_rc)\n",
        "            \n",
        "            # update row/col id -> node dictionary\n",
        "            rc_to_node = dict([i, j] for i, j in enumerate(nodes))\n",
        "        # set tree to root\n",
        "        self.tree = nodes[0]\n",
        "\n",
        "    def update_distance(\n",
        "        self, dist_matrix: np.ndarray, nodes: list, taxa_to_rc: dict\n",
        "    ) -> np.ndarray:\n",
        "        \"\"\"\n",
        "        Method to make a new distance matrix with newer node list.\n",
        "\n",
        "        Parameters\n",
        "        ----------\n",
        "        dist_matrix : np.ndarray of pairwise distances for all taxa\n",
        "        nodes : list of updated nodes\n",
        "        taxa_to_rc : dict for taxa -> row/col id\n",
        "\n",
        "        Returns np.ndarray of pairwise distances for updated nodes\n",
        "        \"\"\"\n",
        "        # dictionary for node -> row/col id\n",
        "        node_to_rc = dict([i, j] for j, i in enumerate(nodes))\n",
        "        \n",
        "        rc = len(nodes)\n",
        "        new_dist_matrix = np.zeros((rc, rc), dtype=float)\n",
        "        for node1 in nodes:\n",
        "            row = node_to_rc[node1]\n",
        "            for node2 in nodes:\n",
        "                node_pairs = list(itertools.product(node1.leaves(), node2.leaves()))\n",
        "                col = node_to_rc[node2]\n",
        "                new_dist_matrix[row, col] = sum(\n",
        "                    dist_matrix[taxa_to_rc[i], taxa_to_rc[j]] for i, j in node_pairs\n",
        "                ) / len(node_pairs)\n",
        "        np.fill_diagonal(new_dist_matrix, np.inf)\n",
        "        return new_dist_matrix\n",
        "\n",
        "\n",
        "def tree_to_newick(t) -> str:\n",
        "    \"\"\"\n",
        "    Function to convert tree to Newick, slightly modified form of the tree.py version.\n",
        "    Takes the root node of an UPGMA tree as input\n",
        "    \"\"\"\n",
        "    if t.right == None:\n",
        "        return t.left + \":\" + str(t.uh)\n",
        "    else:\n",
        "        return (\n",
        "            \"(\"\n",
        "            + \",\".join([tree_to_newick(x) for x in [t.left, t.right]])\n",
        "            + \"):\"\n",
        "            + str(t.uh)\n",
        "        )\n",
        "\n",
        "\n",
        "#main\n",
        "\n",
        "def main ():\n",
        "#to help resolve indent errors \n",
        "    file_in ='./Pdistance.txt'\n",
        "    file_in_2='./Ptaxa.txt'\n",
        "    file_out=\"./Pnewick.txt\"\n",
        "    if os.access(file_in, os.R_OK):\n",
        "        print (\"File \", file_in ,\" is accessible to read\")\n",
        "    else:\n",
        "        print (\"File \", file_in ,\" is not accessible to read, exiting.\")\n",
        "        exit (1)\n",
        "    reader = csv.reader(open(file_in,\"r+\"), delimiter=',')\n",
        "    distance = list(reader)\n",
        "    print(distance)\n",
        "    outer_out_list = []\n",
        "    for inner_list in distance:\n",
        "        innet_out_list = []\n",
        "        for string in inner_list:\n",
        "            innet_out_list.append(float (string))\n",
        "        outer_out_list.append(innet_out_list)\n",
        "    print(outer_out_list)\n",
        "    \n",
        "    if os.access(file_in_2, os.R_OK):\n",
        "        print (\"File \", file_in_2 ,\" is accessible to read\\n\")\n",
        "    else:\n",
        "        print (\"File \", file_in_2 ,\" is not accessible to read, exiting.\")\n",
        "        exit (1)    \n",
        "    reader_2 = csv.reader(open(file_in_2,\"r+\"), delimiter=',')\n",
        "    taxa = list(reader_2)\n",
        "    taxa = taxa[0]\n",
        "    print( taxa )\n",
        "    \n",
        "    distances = np.array( outer_out_list )\n",
        "    x = UPGMA(distances, taxa).tree\n",
        "    final_output_newick=(tree_to_newick(x))\n",
        "    print(final_output_newick)\n",
        "    with open(file_out,\"w\") as f_out:\n",
        "      f_out.write(final_output_newick)\n",
        "    \n",
        "if __name__ == \"__main__\":\n",
        "    main()"
      ],
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "File  ./Pdistance.txt  is accessible to read\n",
            "[['0', '0', '0', '0', '0', '0', '0', '0', '0', '0'], ['3.1818181818181817', '0', '0', '0', '0', '0', '0', '0', '0', '0'], ['2.5462962962962963', '2.2', '0', '0', '0', '0', '0', '0', '0', '0'], ['3.036363636363636', '4.7272727272727275', '2.1636363636363636', '0', '0', '0', '0', '0', '0', '0'], ['3.109090909090909', '4.236363636363636', '2.0727272727272728', '4.390909090909091', '0', '0', '0', '0', '0', '0'], ['3.185185185185185', '4.372727272727273', '2.111111111111111', '4.4', '4.1454545454545455', '0', '0', '0', '0', '0'], ['3.149532710280374', '4.3', '2.259259259259259', '4.209090909090909', '3.9727272727272727', '4.592592592592593', '0', '0', '0', '0'], ['3.0', '4.6909090909090905', '2.2', '5.290909090909091', '4.363636363636363', '4.372727272727273', '4.172727272727273', '0', '0', '0'], ['2.330275229357798', '3.1545454545454548', '1.871559633027523', '3.2636363636363637', '3.081818181818182', '2.9541284403669725', '2.669724770642202', '3.2363636363636363', '0', '0'], ['2.790909090909091', '4.118181818181818', '2.0636363636363635', '4.3545454545454545', '4.0181818181818185', '3.8181818181818183', '3.7454545454545456', '4.327272727272727', '3.018181818181818', '0']]\n",
            "[[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [3.1818181818181817, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [2.5462962962962963, 2.2, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [3.036363636363636, 4.7272727272727275, 2.1636363636363636, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [3.109090909090909, 4.236363636363636, 2.0727272727272728, 4.390909090909091, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [3.185185185185185, 4.372727272727273, 2.111111111111111, 4.4, 4.1454545454545455, 0.0, 0.0, 0.0, 0.0, 0.0], [3.149532710280374, 4.3, 2.259259259259259, 4.209090909090909, 3.9727272727272727, 4.592592592592593, 0.0, 0.0, 0.0, 0.0], [3.0, 4.6909090909090905, 2.2, 5.290909090909091, 4.363636363636363, 4.372727272727273, 4.172727272727273, 0.0, 0.0, 0.0], [2.330275229357798, 3.1545454545454548, 1.871559633027523, 3.2636363636363637, 3.081818181818182, 2.9541284403669725, 2.669724770642202, 3.2363636363636363, 0.0, 0.0], [2.790909090909091, 4.118181818181818, 2.0636363636363635, 4.3545454545454545, 4.0181818181818185, 3.8181818181818183, 3.7454545454545456, 4.327272727272727, 3.018181818181818, 0.0]]\n",
            "File  ./Ptaxa.txt  is accessible to read\n",
            "\n",
            "['NP_990553.1', 'NP_001123565.1', 'ATN38177.1', 'AAP35454.1', 'NP_062002.1', 'NP_001103242.1', 'NP_776351.2', 'NP_001008996.1', 'AAA40590.1', 'AAA19033.1']\n",
            "(((NP_001008996.1:0.0,NP_776351.2:0.0):0.0,(NP_001103242.1:0.0,NP_062002.1:0.0):0.0):0.580231790380873,((AAA19033.1:0.0,AAA40590.1:0.0):0.0,((AAP35454.1:0.0,ATN38177.1:0.0):0.0,(NP_001123565.1:0.0,NP_990553.1:0.0):0.0):0.0):0.580231790380873):0.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "n9Z-Y6SMH4d1"
      },
      "source": [
        "And we are Done!!! "
      ]
    }
  ]
}